<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2023-07-19" />

<title>Bioc2023: extraChIPs Package Demonstration</title>

<script src="index_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="index_files/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="index_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="index_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="index_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
</style>
<script src="index_files/navigation-1.1/tabsets.js"></script>
<link href="index_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="index_files/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
  p.abstract{
    text-align: center;
    font-weight: bold;
  }
  div.abstract{
    margin: auto;
    width: 90%;
  }
</style>


<style type="text/css">
/* for pandoc --citeproc since 2.11 */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="/home/steviep/R/x86_64-pc-linux-gnu-library/4.3/BiocStyle/resources/html/bioconductor.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 828px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




<script>
function toggle_visibility(id1) {
  var e = document.getElementById(id1);
  e.style.display = ((e.style.display!="none") ? "none" : "block");
}
</script>

</head>

<body>


<div class="container-fluid main-container">




<div id="header">



<h1 class="title toc-ignore">Bioc2023: extraChIPs Package Demonstration</h1>
<p class="author-name">Stevie Pederson<span class="affil-mark">1,2,3*</span></p>
<p class="author-affiliation"><span class="affil-mark">1</span>Black Ochre Data Laboratories, Telethon Kids Institute, Adelaide, Australia<br><span class="affil-mark">2</span>Dame Roma Mitchell Cancer Researc Laboratories, University of Adelaide<br><span class="affil-mark">3</span>John Curtin School of Medical Research, Australian National University</p>
<p class="author-email"><span class="affil-mark">*</span><a href="mailto:stephen.pederson.au@gmail.com">stephen.pederson.au@gmail.com</a></p>
<h4 class="date">19 July 2023</h4>
<h4 class="abstract">Abstract</h4>
<p>extraChIPs: Package Demo</p>
<h4 class="package">Package</h4>
<p>extraChIPs 1.5.8</p>

</div>

<h1>Contents</h1>
<div id="TOC">
<ul>
<li><a href="#introduction" id="toc-introduction"><span class="toc-section-number">1</span> Introduction</a>
<ul>
<li><a href="#package-introduction" id="toc-package-introduction"><span class="toc-section-number">1.1</span> Package Introduction</a></li>
<li><a href="#data-and-workshop-setup" id="toc-data-and-workshop-setup"><span class="toc-section-number">1.2</span> Data and Workshop Setup</a></li>
</ul></li>
<li><a href="#core-infrastructure" id="toc-core-infrastructure"><span class="toc-section-number">2</span> Core Infrastructure</a>
<ul>
<li><a href="#mc-functions" id="toc-mc-functions"><span class="toc-section-number">2.1</span> <code>*MC()</code> Functions</a></li>
<li><a href="#tibble-coercion" id="toc-tibble-coercion"><span class="toc-section-number">2.2</span> <code>tibble</code> Coercion</a></li>
</ul></li>
<li><a href="#working-with-peaks" id="toc-working-with-peaks"><span class="toc-section-number">3</span> Working With Peaks</a>
<ul>
<li><a href="#import-and-visualisation" id="toc-import-and-visualisation"><span class="toc-section-number">3.1</span> Import and Visualisation</a></li>
<li><a href="#consensus-peaks" id="toc-consensus-peaks"><span class="toc-section-number">3.2</span> Consensus Peaks</a></li>
<li><a href="#mapping-peaks-to-regulatory-regions" id="toc-mapping-peaks-to-regulatory-regions"><span class="toc-section-number">3.3</span> Mapping Peaks to Regulatory Regions</a></li>
</ul></li>
<li><a href="#differential-signal-detection" id="toc-differential-signal-detection"><span class="toc-section-number">4</span> Differential Signal Detection</a>
<ul>
<li><a href="#using-fixed-width-windows" id="toc-using-fixed-width-windows"><span class="toc-section-number">4.1</span> Using Fixed-Width Windows</a>
<ul>
<li><a href="#reading-in-counts" id="toc-reading-in-counts"><span class="toc-section-number">4.1.1</span> Reading In Counts</a></li>
<li><a href="#visualising-data" id="toc-visualising-data"><span class="toc-section-number">4.1.2</span> Visualising Data</a></li>
<li><a href="#differential-signal-analysis" id="toc-differential-signal-analysis"><span class="toc-section-number">4.1.3</span> Differential Signal Analysis</a></li>
<li><a href="#visualising-results" id="toc-visualising-results"><span class="toc-section-number">4.1.4</span> Visualising Results</a></li>
</ul></li>
<li><a href="#using-sliding-windows" id="toc-using-sliding-windows"><span class="toc-section-number">4.2</span> Using Sliding Windows</a>
<ul>
<li><a href="#defining-windows" id="toc-defining-windows"><span class="toc-section-number">4.2.1</span> Defining Windows</a></li>
<li><a href="#merging-windows" id="toc-merging-windows"><span class="toc-section-number">4.2.2</span> Merging Windows</a></li>
</ul></li>
</ul></li>
<li><a href="#comparing-both-sets-of-results" id="toc-comparing-both-sets-of-results"><span class="toc-section-number">5</span> Comparing Both Sets Of Results</a>
<ul>
<li><a href="#association-between-datasets" id="toc-association-between-datasets"><span class="toc-section-number">5.1</span> Association Between Datasets</a></li>
<li><a href="#regions-in-context" id="toc-regions-in-context"><span class="toc-section-number">5.2</span> Regions in Context</a></li>
<li><a href="#profile-heatmaps" id="toc-profile-heatmaps"><span class="toc-section-number">5.3</span> Profile Heatmaps</a></li>
</ul></li>
<li><a href="#references" id="toc-references">References</a></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li><a href="#session-info" id="toc-session-info"><span class="toc-section-number">A</span> Session info</a></li>
</ul>
</div>

<div id="introduction" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<div id="package-introduction" class="section level2" number="1.1">
<h2><span class="header-section-number">1.1</span> Package Introduction</h2>
<p><code>extraChIPs</code> was developed to provide key infrastructure for a larger ChIP-Seq workflow comparing binding dynamics across multiple ChIP targets.
The focus is on:</p>
<ul>
<li>Differential Signal Detection</li>
<li>Comparison between ChIP targets or sets of results</li>
<li>Working with <code>GRanges</code> objects containing important values in the <code>mcols()</code> element</li>
<li>Visualisation of Data and Results</li>
</ul>
<p>The package was developed to utilise <em>existing</em> S4 Object Classes and behave in a <em>tidy-friendly manner</em>, whilst not being strictly designed under tidy programming guidelines.
Some functions may be analogous to those in existing packages, but the aim is to be simple and intuitive to use, whilst simplifying high-performance analyses.
Previous work by the authors of DiffBind <span class="citation">(<a href="#ref-DiffBind2012">Ross-Innes et al. 2012</a>)</span> and csaw <span class="citation">(<a href="#ref-csaw2016">Lun and Smyth 2016</a>)</span> has been instrumental in the package design, along with the tidyverse <span class="citation">(<a href="#ref-tidyverse2019">Wickham et al. 2019</a>)</span> and edgeR <span class="citation">(<a href="#ref-edgeRQLF2016">Chen, Lun, and Smyth 2016</a>)</span>.
The intention of the package is to integrate easily with these packages, wrap multiple functions where appropriate, and provide new tools for both visualisation and working with results.</p>
<p><strong>The talk accompanying this material can be found</strong> <a href="Bioc2023_presentation.html">here</a></p>
</div>
<div id="data-and-workshop-setup" class="section level2" number="1.2">
<h2><span class="header-section-number">1.2</span> Data and Workshop Setup</h2>
<p>Today’s workshop will focus on the <code>devel</code> version of the package, which can be installed using</p>
<pre class="r"><code>BiocManager::install(
  &quot;extraChIPs&quot;, version = &quot;devel&quot;, update = FALSE, ask = FALSE
)</code></pre>
<p>Alternatively, this can be installed directly from github</p>
<pre class="r"><code>BiocManager::install(
  &quot;smped/extraChIPs&quot;, ref = &quot;devel&quot;, update = FALSE, ask = FALSE
)</code></pre>
<p>Additional packages required can be installed as follows:</p>
<pre class="r"><code>pkg &lt;- c(
  &quot;rtracklayer&quot;, &quot;tidyverse&quot;, &quot;glue&quot;, &quot;plyranges&quot;, &quot;Rsamtools&quot;, &quot;csaw&quot;, 
  &quot;edgeR&quot;,&quot;patchwork&quot;
)
BiocManager::install(pkgs = pkg, update = FALSE, ask = FALSE)</code></pre>
<pre class="r"><code>library(tidyverse)
library(glue)
library(plyranges)
library(extraChIPs)
library(Rsamtools)
library(csaw)
library(edgeR)
library(rtracklayer)
library(patchwork)
theme_set(theme_bw())</code></pre>
<p>Data can be obtained from <a href="https://github.com/smped/extraChIPs_demo_data">here</a> and should be placed in a folder called <code>data</code> within your current working directory.
This data has been pre-processed <a href="https://github.com/smped/PRJNA509779">here</a> using the prepareChIPs workflow <span class="citation">(<a href="#ref-pederson_2023">Pederson 2023</a>)</span>, and subset to a region on Chromosome 10 for easier working in today’s context.</p>
<p>The two targets in this dataset are the Estrogen Receptor (ER<span class="math inline">\(\alpha\)</span>) and the histone mark H3K27ac.
Data has been taken from ZR-75-1 cells cultured in Estrogen (E2) or Estrogen + Dihydrotestosterone (E2DHT), and was first published by researchers from the Dame Roma Mitchell Cancer Research Laboratories <span class="citation">(<a href="#ref-Hickey2021-mz">Hickey et al. 2021</a>)</span></p>
<p>The sample sheet for the ER<span class="math inline">\(\alpha\)</span> samples can be setup by copying the following code.
As can be seen, we have three samples from the E2 (baseline) group with three more treated by the addition of DHT.</p>
<pre class="r"><code>treat_levels &lt;- c(&quot;E2&quot;, &quot;E2DHT&quot;)
er_samples &lt;- tibble(
  accession = paste0(&quot;SRR831518&quot;, seq(0, 5)),
  target = &quot;ER&quot;,
  treatment = factor(rep(treat_levels, each = 3), levels = treat_levels)
)
er_samples</code></pre>
<pre><code>## # A tibble: 6 × 3
##   accession  target treatment
##   &lt;chr&gt;      &lt;chr&gt;  &lt;fct&gt;    
## 1 SRR8315180 ER     E2       
## 2 SRR8315181 ER     E2       
## 3 SRR8315182 ER     E2       
## 4 SRR8315183 ER     E2DHT    
## 5 SRR8315184 ER     E2DHT    
## 6 SRR8315185 ER     E2DHT</code></pre>
<p>We’ll also setup colours for each treatment for consistent visualisations throughout the analysis</p>
<pre class="r"><code>treat_colours &lt;- setNames(c(&quot;steelblue&quot;, &quot;red3&quot;), treat_levels)</code></pre>
<p>Finally, we’ll define a <code>Seqinfo</code> object just for today’s data.</p>
<pre class="r"><code>sq &lt;- Seqinfo(
  seqnames = &quot;chr10&quot;, seqlengths = 135534747, isCircular = FALSE, 
  genome = &quot;hg19&quot;
)</code></pre>
</div>
</div>
<div id="core-infrastructure" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Core Infrastructure</h1>
<div id="mc-functions" class="section level2" number="2.1">
<h2><span class="header-section-number">2.1</span> <code>*MC()</code> Functions</h2>
<p>Many functions exist for combining sets of ranges, such as <code>reduce()</code>, <code>intersect()</code>, <code>setdiff()</code> etc, however these have been extended in <code>extraChIPs</code> to <code>reduceMC()</code>, <code>intersectMC()</code>, <code>setdiffMC()</code> and <code>unionMC()</code> as well as tidy-aligned functions <code>chopMC()</code> and <code>distinctMC()</code>.
This group of functions behaves identically to the core <code>GenomicRanges</code> versions, but retains data in the <code>mcols</code> element.
These are used internally by many functions within the package, but also have many stand-alone uses.</p>
<p>Define a simple set of ranges, with ids in the <code>mcols</code></p>
<pre class="r"><code>x &lt;- GRanges(c(&quot;chr1:1-10&quot;, &quot;chr1:6-12&quot;))
x$id &lt;- c(&quot;range1&quot;, &quot;range2&quot;)
x$gene &lt;- &quot;gene1&quot;
x</code></pre>
<pre><code>## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames    ranges strand |          id        gene
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;
##   [1]     chr1      1-10      * |      range1       gene1
##   [2]     chr1      6-12      * |      range2       gene1
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>When calling <code>reduce()</code> we lose all <code>mcols</code> information.</p>
<pre class="r"><code>reduce(x)</code></pre>
<pre><code>## GRanges object with 1 range and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1      1-12      *
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>These are retained when calling <code>reduceMC</code></p>
<pre class="r"><code>reduceMC(x)</code></pre>
<pre><code>## GRanges object with 1 range and 2 metadata columns:
##       seqnames    ranges strand |              id        gene
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;CharacterList&gt; &lt;character&gt;
##   [1]     chr1      1-12      * |   range1,range2       gene1
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>All other functions behave in a similar manner</p>
<pre class="r"><code>y &lt;- GRanges(c(&quot;chr1:11-15&quot;))
intersect(x, y)</code></pre>
<pre><code>## GRanges object with 1 range and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1     11-12      *
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<pre class="r"><code>intersectMC(x, y)</code></pre>
<pre><code>## GRanges object with 1 range and 2 metadata columns:
##       seqnames    ranges strand |          id        gene
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;
##   [1]     chr1     11-12      * |      range2       gene1
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<pre class="r"><code>setdiff(x, y)</code></pre>
<pre><code>## GRanges object with 1 range and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1      1-10      *
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<pre class="r"><code>setdiffMC(x, y)</code></pre>
<pre><code>## GRanges object with 1 range and 2 metadata columns:
##       seqnames    ranges strand |              id        gene
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;CharacterList&gt; &lt;character&gt;
##   [1]     chr1      1-10      * |   range1,range2       gene1
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<pre class="r"><code>union(x, y)</code></pre>
<pre><code>## GRanges object with 1 range and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1      1-15      *
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<pre class="r"><code>unionMC(x, y)</code></pre>
<pre><code>## GRanges object with 1 range and 2 metadata columns:
##       seqnames    ranges strand |              id        gene
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;CharacterList&gt; &lt;character&gt;
##   [1]     chr1      1-15      * |   range1,range2       gene1
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="tibble-coercion" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> <code>tibble</code> Coercion</h2>
<p>As this package is also designed to play well with the <code>tidyverse</code> the functions <code>as_tibble()</code> and <code>colToRanges()</code> have been introduced.
When coercing to a <code>tibble</code>, the <code>GRanges</code> element is coerced to a character vector.</p>
<pre class="r"><code>tbl &lt;- as_tibble(x)
tbl</code></pre>
<pre><code>## # A tibble: 2 × 3
##   range     id     gene 
##   &lt;chr&gt;     &lt;chr&gt;  &lt;chr&gt;
## 1 chr1:1-10 range1 gene1
## 2 chr1:6-12 range2 gene1</code></pre>
<p>The tibble can be returned to the more conventional set of columns by setting <code>rangeAsChar = FALSE</code></p>
<pre class="r"><code>as_tibble(x, rangeAsChar = FALSE)</code></pre>
<pre><code>## # A tibble: 2 × 7
##   seqnames start   end width strand id     gene 
##   &lt;fct&gt;    &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;fct&gt;  &lt;chr&gt;  &lt;chr&gt;
## 1 chr1         1    10    10 *      range1 gene1
## 2 chr1         6    12     7 *      range2 gene1</code></pre>
<p>Additional methods have been implemented for easy coercion of <code>DataFrame</code>, <code>GInteractions</code>, <code>Seqinfo</code>, <code>SummarzedExperiment</code> and <code>TopTags</code> objects</p>
<p>Coercion is implemented from <code>tibble</code> to <code>GRanges</code> using <code>colToRanges()</code></p>
<pre class="r"><code>colToRanges(tbl, var = &quot;range&quot;)</code></pre>
<pre><code>## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames    ranges strand |          id        gene
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;
##   [1]     chr1      1-10      * |      range1       gene1
##   [2]     chr1      6-12      * |      range2       gene1
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>Whilst the <code>seqinfo</code> is clearly lost in the reverse coercion, <code>colToRanges()</code> also allows for moving a <code>GRanges</code> object within the <code>mcols</code> element to be set as a new set of ranges.
In this case, the <code>seqinfo</code> will be retained.</p>
</div>
</div>
<div id="working-with-peaks" class="section level1" number="3">
<h1><span class="header-section-number">3</span> Working With Peaks</h1>
<div id="import-and-visualisation" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> Import and Visualisation</h2>
<p>Broad and Narrow Peak files as output by peak callers such as <code>macs2 callpeak</code> <span class="citation">(<a href="#ref-Zhang2008-ms">Zhang et al. 2008</a>)</span> can be imported using <code>importPeaks()</code>.
This will always return a <code>GRangesList</code> for simple downstream analysis.
If importing <code>narrowPeak</code> files, the peak centre can be automatically added.</p>
<pre class="r"><code>fl &lt;- glue(&quot;data/ER/{er_samples$accession}_peaks.narrowPeak&quot;)
peaks &lt;- importPeaks(fl, seqinfo = sq, centre = TRUE)
names(peaks) &lt;- str_remove_all(names(peaks), &quot;_peaks.narrowPeak&quot;)</code></pre>
<p>(Names can also be set using <code>glue</code> syntax, however the above is generally more easily comprehended at first glance.)</p>
<p>Blacklisted regions and grey-listed regions are also able to be applied during import.
Blacklisted regions are usually available from <a href="https://github.com/Boyle-Lab/Blacklist/tree/master/lists">public sources</a>, whilst grey-listed regions are derived from a relevant input.</p>
<p>These have also been provided and can be loaded separately then combined.</p>
<pre class="r"><code>bg_list &lt;- file.path(
  &quot;data&quot;, &quot;annotations&quot;, c(&quot;chr10_blacklist.bed&quot;, &quot;chr10_greylist.bed&quot;)
) %&gt;% 
  lapply(import.bed, seqinfo = sq) %&gt;% 
  lapply(granges) %&gt;% 
  GRangesList() %&gt;% 
  unlist() %&gt;% 
  sort()
peaks &lt;- importPeaks(fl, seqinfo = sq, centre = TRUE, blacklist = bg_list)
names(peaks) &lt;- str_remove_all(names(peaks), &quot;_peaks.narrowPeak&quot;)</code></pre>
<p>Now we’ve loaded our peaks and excluded peaks within unreliable ranges, one of the first tasks might be to check how similar our replicates are.
This can easily be performed using <code>plotOverlaps()</code></p>
<pre class="r"><code>plotOverlaps(peaks, set_col = treat_colours[er_samples$treatment])</code></pre>
<div class="figure"><span style="display:block;" id="fig:plot-overlaps"></span>
<img src="index_files/figure-html/plot-overlaps-1.png" alt="Overlapping peaks between all replicates" width="100%" />
<p class="caption">
Figure 1: <span class="caption-title">Overlapping peaks between all replicates</span><br>
</p>
</div>
<p><code>plotOverlaps()</code> calls the plotting functions provided by <code>ComplexUpset::upset()</code> <span class="citation">(<a href="#ref-ComplexUpset2020">Krassowski 2020</a>)</span> and as such, can handle any additional parameters which are understood by this function.</p>
<pre class="r"><code>plotOverlaps(
  peaks, .sort_sets = FALSE, 
  set_col = treat_colours[er_samples$treatment], n_intersections = 10
)</code></pre>
<div class="figure"><span style="display:block;" id="fig:plot-overlaps2"></span>
<img src="index_files/figure-html/plot-overlaps2-1.png" alt="Overlapping peaks, showing only intersections with 10 or more members and retaining the original sample ordering" width="100%" />
<p class="caption">
Figure 2: <span class="caption-title">Overlapping peaks, showing only intersections with 10 or more members and retaining the original sample ordering</span><br>
</p>
</div>
</div>
<div id="consensus-peaks" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> Consensus Peaks</h2>
<p>The next step might be to form a set of consensus peaks for each treatment.
Just forming a set of consensus peaks for the E2 samples can be done manually</p>
<pre class="r"><code>makeConsensus(peaks[1:3])</code></pre>
<pre><code>## GRanges object with 244 ranges and 4 metadata columns:
##         seqnames            ranges strand | SRR8315180 SRR8315181 SRR8315182
##            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;logical&gt;  &lt;logical&gt;  &lt;logical&gt;
##     [1]    chr10 43048195-43048529      * |       TRUE       TRUE       TRUE
##     [2]    chr10 43521739-43522260      * |       TRUE       TRUE       TRUE
##     [3]    chr10 43540042-43540390      * |       TRUE      FALSE       TRUE
##     [4]    chr10 43606238-43606573      * |       TRUE       TRUE       TRUE
##     [5]    chr10 43851214-43851989      * |      FALSE       TRUE       TRUE
##     ...      ...               ...    ... .        ...        ...        ...
##   [240]    chr10 99168353-99168649      * |       TRUE       TRUE       TRUE
##   [241]    chr10 99207868-99208156      * |      FALSE       TRUE       TRUE
##   [242]    chr10 99326100-99326363      * |      FALSE      FALSE       TRUE
##   [243]    chr10 99331363-99331730      * |       TRUE       TRUE       TRUE
##   [244]    chr10 99621632-99621961      * |      FALSE       TRUE       TRUE
##                 n
##         &lt;numeric&gt;
##     [1]         3
##     [2]         3
##     [3]         2
##     [4]         3
##     [5]         2
##     ...       ...
##   [240]         3
##   [241]         2
##   [242]         1
##   [243]         3
##   [244]         2
##   -------
##   seqinfo: 1 sequence from hg19 genome</code></pre>
<p>In reality, we may prefer to only retain peaks found in 2/3 samples and we can do this by setting <code>p = 2/3</code>.
We can also also <code>makeConsensus</code> to retain one or more columns from the original peaks using <code>var =</code></p>
<pre class="r"><code>makeConsensus(peaks[1:3], p = 2/3, var = &quot;centre&quot;)</code></pre>
<pre><code>## GRanges object with 164 ranges and 5 metadata columns:
##         seqnames            ranges strand |                     centre
##            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |              &lt;NumericList&gt;
##     [1]    chr10 43048195-43048529      * | 43048341,43048357,43048389
##     [2]    chr10 43521739-43522260      * | 43522008,43522012,43522039
##     [3]    chr10 43540042-43540390      * |          43540298,43540245
##     [4]    chr10 43606238-43606573      * | 43606418,43606408,43606421
##     [5]    chr10 43851214-43851989      * |          43851672,43851766
##     ...      ...               ...    ... .                        ...
##   [160]    chr10 99096784-99097428      * | 99097259,99097253,99097249
##   [161]    chr10 99168353-99168649      * | 99168506,99168486,99168513
##   [162]    chr10 99207868-99208156      * |          99207995,99208002
##   [163]    chr10 99331363-99331730      * | 99331621,99331585,99331580
##   [164]    chr10 99621632-99621961      * |          99621809,99621828
##         SRR8315180 SRR8315181 SRR8315182         n
##          &lt;logical&gt;  &lt;logical&gt;  &lt;logical&gt; &lt;numeric&gt;
##     [1]       TRUE       TRUE       TRUE         3
##     [2]       TRUE       TRUE       TRUE         3
##     [3]       TRUE      FALSE       TRUE         2
##     [4]       TRUE       TRUE       TRUE         3
##     [5]      FALSE       TRUE       TRUE         2
##     ...        ...        ...        ...       ...
##   [160]       TRUE       TRUE       TRUE         3
##   [161]       TRUE       TRUE       TRUE         3
##   [162]      FALSE       TRUE       TRUE         2
##   [163]       TRUE       TRUE       TRUE         3
##   [164]      FALSE       TRUE       TRUE         2
##   -------
##   seqinfo: 1 sequence from hg19 genome</code></pre>
<p>By default, this takes the union of the ranges called as peaks in each sample.
We may instead prefer to just retain regions <em>covered</em> in <span class="math inline">\(\geq\)</span> 2/3 samples.
Not that whilst we have the same number of ranges here, the ranges are tighter.</p>
<pre class="r"><code>makeConsensus(peaks[1:3], p = 2/3, method = &quot;coverage&quot;)</code></pre>
<pre><code>## GRanges object with 164 ranges and 4 metadata columns:
##         seqnames            ranges strand | SRR8315180 SRR8315181 SRR8315182
##            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;logical&gt;  &lt;logical&gt;  &lt;logical&gt;
##     [1]    chr10 43048224-43048519      * |       TRUE       TRUE       TRUE
##     [2]    chr10 43521773-43522218      * |       TRUE       TRUE       TRUE
##     [3]    chr10 43540200-43540384      * |       TRUE      FALSE       TRUE
##     [4]    chr10 43606264-43606559      * |       TRUE       TRUE       TRUE
##     [5]    chr10 43851570-43851940      * |      FALSE       TRUE       TRUE
##     ...      ...               ...    ... .        ...        ...        ...
##   [160]    chr10 99097038-99097398      * |       TRUE       TRUE       TRUE
##   [161]    chr10 99168367-99168608      * |       TRUE       TRUE       TRUE
##   [162]    chr10 99207908-99208139      * |      FALSE       TRUE       TRUE
##   [163]    chr10 99331412-99331687      * |       TRUE       TRUE       TRUE
##   [164]    chr10 99621674-99621944      * |      FALSE       TRUE       TRUE
##                 n
##         &lt;numeric&gt;
##     [1]         3
##     [2]         3
##     [3]         2
##     [4]         3
##     [5]         2
##     ...       ...
##   [160]         3
##   [161]         3
##   [162]         2
##   [163]         3
##   [164]         2
##   -------
##   seqinfo: 1 sequence from hg19 genome</code></pre>
<p>In order to make consensus peaks for each condition, we can split by treatment group, then use a simple <code>lapply()</code> trick, to work with both treatment groups together.</p>
<pre class="r"><code>cons_peaks &lt;- peaks %&gt;% 
  split(f = er_samples$treatment) %&gt;% 
  lapply(makeConsensus, p = 2/3, var = &quot;centre&quot;) %&gt;% 
  lapply(select, centre) %&gt;% 
  lapply(mutate, centre = vapply(centre, mean, numeric(1))) %&gt;% 
  GRangesList()</code></pre>
<p>A lot happened in that code chunk, so let’s break it down</p>
<ol style="list-style-type: decimal">
<li>We split samples based on the treatment (<code>split(f = er_samples$treatment)</code>), then</li>
<li>Identified consensus peaks within each treatment found in 2/3 samples, retaining the peak centre (<code>lapply(makeConsensus, p = 2/3, var = "centre")</code>)</li>
<li>Subset the <code>mcols</code> to only retain the peak centre (<code>lapply(select, centre)</code>).</li>
<li>The previous step returned centre estimates as a <code>NumericList</code>, so we took the mean of these for each consensus peak</li>
<li>We coerced to a <code>GRangeList</code></li>
</ol>
<p>Now we can compare our two treatment conditions using <code>plotOverlaps()</code> again, which will return a Venn Diagram when passing a <code>GRangesList</code> of length 2.</p>
<pre class="r"><code>plotOverlaps(cons_peaks, set_col = treat_colours)</code></pre>
<div class="figure"><span style="display:block;" id="fig:plot-cons-overlap"></span>
<img src="index_files/figure-html/plot-cons-overlap-1.png" alt="Overlaps betwen consensus peaks specific to each treatment group" width="100%" />
<p class="caption">
Figure 3: <span class="caption-title">Overlaps betwen consensus peaks specific to each treatment group</span><br>
</p>
</div>
<p>Finally, we can form a set of “union” peaks, defined as those within a consensus peak for either treatment group.</p>
<pre class="r"><code>union_peaks &lt;- makeConsensus(cons_peaks, var = &quot;centre&quot;) %&gt;% 
  mutate(centre = vapply(centre, mean, numeric(1)) %&gt;% round(0))</code></pre>
</div>
<div id="mapping-peaks-to-regulatory-regions" class="section level2" number="3.3">
<h2><span class="header-section-number">3.3</span> Mapping Peaks to Regulatory Regions</h2>
<p>Now we have detected our ranges where ER<span class="math inline">\(\alpha\)</span> appears to have bound, we might wish to find which genes are likely targets and what type of genomic feature is associated with each range.
First let’s define some genomic regions.</p>
<p>A Gencode GTF has been provided in data, restricted to the regions of chr10 we’re working with today.
We can load this then, split based on the feature type, giving a <code>GRangesList</code> with gene, transcript and exon-level features.</p>
<pre class="r"><code>gtf &lt;- file.path(
  &quot;data&quot;, &quot;annotations&quot;, &quot;gencode.v43lift37.chr10.annotation.gtf.gz&quot;
) %&gt;% 
  import.gff() %&gt;% 
  split(.$type)
seqinfo(gtf) &lt;- sq</code></pre>
<p>The function <code>defineRegions()</code> enables us to define 1) Promoter, 2) Upstream Promoter, 3) Intron, 4) Exon, 5) Proximal Intergenic and 6) Distal Intergenic Regions.
The distances for these regions can be specified by the user, but here we’ll just use the defaults.</p>
<pre class="r"><code>regions &lt;- defineRegions(
  genes = gtf$gene, transcripts = gtf$transcript, exons = gtf$exon
)
regions</code></pre>
<pre><code>## GRangesList object of length 6:
## $promoter
## GRanges object with 3359 ranges and 5 metadata columns:
##          seqnames              ranges strand |                region
##             &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |           &lt;character&gt;
##      [1]    chr10         61989-64988      * | Promoter (-2500/+500)
##      [2]    chr10         66360-69392      * | Promoter (-2500/+500)
##      [3]    chr10         88152-91151      * | Promoter (-2500/+500)
##      [4]    chr10         94710-97961      * | Promoter (-2500/+500)
##      [5]    chr10       119604-122603      * | Promoter (-2500/+500)
##      ...      ...                 ...    ... .                   ...
##   [3355]    chr10 135488075-135491074      * | Promoter (-2500/+500)
##   [3356]    chr10 135491384-135494383      * | Promoter (-2500/+500)
##   [3357]    chr10 135494684-135497683      * | Promoter (-2500/+500)
##   [3358]    chr10 135503135-135506134      * | Promoter (-2500/+500)
##   [3359]    chr10 135513071-135518323      * | Promoter (-2500/+500)
##                                                                  gene_id
##                                                          &lt;CharacterList&gt;
##      [1]                                            ENSG00000260370.2_11
##      [2]  ENSG00000260370.2_11,ENSG00000260370.2_11,ENSG00000260370.2_11
##      [3]                                             ENSG00000237297.1_7
##      [4] ENSG00000261456.6_7,ENSG00000261456.6_7,ENSG00000261456.6_7,...
##      [5]                                             ENSG00000261456.6_7
##      ...                                                             ...
##   [3355]                                             ENSG00000276046.1_5
##   [3356]                                             ENSG00000276904.1_5
##   [3357]                                             ENSG00000278641.1_5
##   [3358]                                               ENSG00000226880.1
##   [3359] ENSG00000213147.3_3,ENSG00000261914.2_3,ENSG00000282875.1_8,...
##                                                gene_name
##                                          &lt;CharacterList&gt;
##      [1]                                 ENSG00000260370
##      [2] ENSG00000260370,ENSG00000260370,ENSG00000260370
##      [3]                                 ENSG00000237297
##      [4]                           TUBB8,TUBB8,TUBB8,...
##      [5]                                           TUBB8
##      ...                                             ...
##   [3355]                                         DUX4L13
##   [3356]                                         DUX4L14
##   [3357]                                         DUX4L15
##   [3358]                                    XX-2136C48.7
##   [3359]         RPL23AP60,RPL23AP84,ENSG00000282875,...
##                                                            transcript_id
##                                                          &lt;CharacterList&gt;
##      [1]                                             ENST00000562162.2_4
##      [2]     ENST00000682078.1_1,ENST00000566940.2_4,ENST00000683798.1_1
##      [3]                                             ENST00000416477.1_2
##      [4] ENST00000568584.6_6,ENST00000568866.5_4,ENST00000561967.1_4,...
##      [5]                                             ENST00000564130.2_4
##      ...                                                             ...
##   [3355]                                             ENST00000554103.2_2
##   [3356]                                             ENST00000621227.1_2
##   [3357]                                             ENST00000611059.1_2
##   [3358]                                               ENST00000411632.1
##   [3359] ENST00000450011.1_2,ENST00000571193.2_2,ENST00000635574.1_4,...
##                                          transcript_name
##                                          &lt;CharacterList&gt;
##      [1]                                 ENST00000562162
##      [2] ENST00000682078,ENST00000566940,ENST00000683798
##      [3]                                 ENST00000416477
##      [4]               TUBB8-206,TUBB8-207,TUBB8-201,...
##      [5]                                       TUBB8-204
##      ...                                             ...
##   [3355]                                     DUX4L13-201
##   [3356]                                     DUX4L14-201
##   [3357]                                     DUX4L15-201
##   [3358]                                XX-2136C48.7-001
##   [3359] RPL23AP60-201,RPL23AP84-201,ENST00000635574,...
##   -------
##   seqinfo: 1 sequence from hg19 genome
## 
## ...
## &lt;5 more elements&gt;</code></pre>
<p>Now we can assign these to the peaks using <code>bestOverlap</code></p>
<pre class="r"><code>union_peaks$region &lt;- bestOverlap(union_peaks, regions)
union_peaks</code></pre>
<pre><code>## GRanges object with 188 ranges and 5 metadata columns:
##         seqnames            ranges strand |    centre        E2     E2DHT
##            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;logical&gt; &lt;logical&gt;
##     [1]    chr10 43048195-43048529      * |  43048372      TRUE      TRUE
##     [2]    chr10 43521739-43522260      * |  43522014      TRUE      TRUE
##     [3]    chr10 43540042-43540457      * |  43540310      TRUE      TRUE
##     [4]    chr10 43606184-43606573      * |  43606417      TRUE      TRUE
##     [5]    chr10 43851214-43851989      * |  43851719      TRUE     FALSE
##     ...      ...               ...    ... .       ...       ...       ...
##   [184]    chr10 99168349-99168690      * |  99168489      TRUE      TRUE
##   [185]    chr10 99207862-99208157      * |  99208004      TRUE      TRUE
##   [186]    chr10 99331323-99331741      * |  99331592      TRUE      TRUE
##   [187]    chr10 99340107-99340354      * |  99340248     FALSE      TRUE
##   [188]    chr10 99621632-99621988      * |  99621798      TRUE      TRUE
##                 n              region
##         &lt;numeric&gt;         &lt;character&gt;
##     [1]         2            promoter
##     [2]         2   distal_intergenic
##     [3]         2   distal_intergenic
##     [4]         2              intron
##     [5]         1 proximal_intergenic
##     ...       ...                 ...
##   [184]         2              intron
##   [185]         2            promoter
##   [186]         2            promoter
##   [187]         1   upstream_promoter
##   [188]         2   upstream_promoter
##   -------
##   seqinfo: 1 sequence from hg19 genome</code></pre>
<p>When providing a <code>GRangesList</code> the overlaps are defined by the element names.
Alternatively, we could pass a <code>GRanges</code> object and specify the name from the <code>mcols</code> element we wish to us.</p>
<pre class="r"><code>union_peaks$region &lt;- bestOverlap(union_peaks, unlist(regions), var = &quot;region&quot;)
union_peaks</code></pre>
<pre><code>## GRanges object with 188 ranges and 5 metadata columns:
##         seqnames            ranges strand |    centre        E2     E2DHT
##            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;logical&gt; &lt;logical&gt;
##     [1]    chr10 43048195-43048529      * |  43048372      TRUE      TRUE
##     [2]    chr10 43521739-43522260      * |  43522014      TRUE      TRUE
##     [3]    chr10 43540042-43540457      * |  43540310      TRUE      TRUE
##     [4]    chr10 43606184-43606573      * |  43606417      TRUE      TRUE
##     [5]    chr10 43851214-43851989      * |  43851719      TRUE     FALSE
##     ...      ...               ...    ... .       ...       ...       ...
##   [184]    chr10 99168349-99168690      * |  99168489      TRUE      TRUE
##   [185]    chr10 99207862-99208157      * |  99208004      TRUE      TRUE
##   [186]    chr10 99331323-99331741      * |  99331592      TRUE      TRUE
##   [187]    chr10 99340107-99340354      * |  99340248     FALSE      TRUE
##   [188]    chr10 99621632-99621988      * |  99621798      TRUE      TRUE
##                 n                 region
##         &lt;numeric&gt;            &lt;character&gt;
##     [1]         2  Promoter (-2500/+500)
##     [2]         2     Intergenic (&gt;10kb)
##     [3]         2     Intergenic (&gt;10kb)
##     [4]         2                 Intron
##     [5]         1     Intergenic (&lt;10kb)
##     ...       ...                    ...
##   [184]         2                 Intron
##   [185]         2  Promoter (-2500/+500)
##   [186]         2  Promoter (-2500/+500)
##   [187]         1 Upstream Promoter (-..
##   [188]         2 Upstream Promoter (-..
##   -------
##   seqinfo: 1 sequence from hg19 genome</code></pre>
<p>Let’s set these as a factor in order as contained in the original set of regions.</p>
<pre class="r"><code>reg_levels &lt;- vapply(regions, function(x) x$region[1], character(1))
union_peaks$region &lt;- factor(union_peaks$region, levels = reg_levels)</code></pre>
<p>We can check the distribution of our peaks against these regions using <code>plotPie()</code></p>
<pre class="r"><code>region_colours &lt;- hcl.colors(6, &quot;viridis&quot;, rev = TRUE)
names(region_colours) &lt;- reg_levels
plotPie(union_peaks, fill = &quot;region&quot;) +
  scale_fill_manual(values = region_colours)</code></pre>
<div class="figure"><span style="display:block;" id="fig:plotpie-default"></span>
<img src="index_files/figure-html/plotpie-default-1.png" alt="Distribution of peaks within defined regulatory regions" width="100%" />
<p class="caption">
Figure 4: <span class="caption-title">Distribution of peaks within defined regulatory regions</span><br>
</p>
</div>
<p>The labels for <code>plotPie()</code> use the <code>glue</code> syntax, lending enormous flexibility to what can be placed here.
Whist the defaults are hopefully useful, in the code below. we’ll modify these to wrap the text using <code>str_wrap()</code> and add <code>n =</code> to the peak numbers.</p>
<pre class="r"><code>plotPie(
  union_peaks, fill = &quot;region&quot;, total_size = 5,
  cat_glue = &quot;{str_wrap(.data[[fill]], 15)}\nn = {comma(n, 1)}\n({percent(p, 0.1)})&quot;
) +
  scale_fill_manual(values = region_colours)</code></pre>
<div class="figure"><span style="display:block;" id="fig:plotpie-labels"></span>
<img src="index_files/figure-html/plotpie-labels-1.png" alt="Distribution of peaks with manually tweaked segment labels" width="100%" />
<p class="caption">
Figure 5: <span class="caption-title">Distribution of peaks with manually tweaked segment labels</span><br>
</p>
</div>
<p>Finally, we might wish to know which genes are likely targets for each peak.
<code>mapByFeature()</code> uses a mapping strategy which maps peaks/ranges to genes differently depending on which type of feature they overlap.</p>
<ol style="list-style-type: decimal">
<li>Ranges overlapping a <em>promoter</em> are only assigned to the gene(s) associated with that promoter</li>
<li>Ranges overlapping an <em>enhancer</em> are mapped to all gene within a given distance</li>
<li>Ranges overlapping a <em>long-range interaction</em> are mapped to genes at both ends of the interaction</li>
<li>Remaining ranges are mapped to the <em>nearest gene</em> within a given distance</li>
</ol>
<pre class="r"><code>union_peaks &lt;- mapByFeature(
  union_peaks, genes = gtf$gene, prom = regions$promoter
)
union_peaks %&gt;% filter(str_detect(region, &quot;^Promoter&quot;), !E2)</code></pre>
<pre><code>## GRanges object with 7 ranges and 7 metadata columns:
##       seqnames            ranges strand |    centre        E2     E2DHT
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;logical&gt; &lt;logical&gt;
##   [1]    chr10 54169026-54169221      * |  54169133     FALSE      TRUE
##   [2]    chr10 63808711-63809004      * |  63808893     FALSE      TRUE
##   [3]    chr10 69847132-69847384      * |  69847254     FALSE      TRUE
##   [4]    chr10 71879941-71880280      * |  71880136     FALSE      TRUE
##   [5]    chr10 74114141-74114500      * |  74114282     FALSE      TRUE
##   [6]    chr10 75579311-75579548      * |  75579425     FALSE      TRUE
##   [7]    chr10 88281479-88281723      * |  88281596     FALSE      TRUE
##               n                region
##       &lt;numeric&gt;              &lt;factor&gt;
##   [1]         1 Promoter (-2500/+500)
##   [2]         1 Promoter (-2500/+500)
##   [3]         1 Promoter (-2500/+500)
##   [4]         1 Promoter (-2500/+500)
##   [5]         1 Promoter (-2500/+500)
##   [6]         1 Promoter (-2500/+500)
##   [7]         1 Promoter (-2500/+500)
##                                          gene_id       gene_name
##                                  &lt;CharacterList&gt; &lt;CharacterList&gt;
##   [1]                        ENSG00000227972.1_6        THAP12P3
##   [2]                      ENSG00000150347.17_12          ARID5B
##   [3]                       ENSG00000138347.17_9            MYPN
##   [4]                       ENSG00000042286.15_8           AIFM2
##   [5]                      ENSG00000148719.16_11         DNAJB12
##   [6]                      ENSG00000148660.22_13          CAMK2G
##   [7] ENSG00000062650.20_11,ENSG00000227896.2_11    WAPL,WAPL-DT
##   -------
##   seqinfo: 1 sequence from hg19 genome</code></pre>
</div>
</div>
<div id="differential-signal-detection" class="section level1" number="4">
<h1><span class="header-section-number">4</span> Differential Signal Detection</h1>
<div id="using-fixed-width-windows" class="section level2" number="4.1">
<h2><span class="header-section-number">4.1</span> Using Fixed-Width Windows</h2>
<div id="reading-in-counts" class="section level3" number="4.1.1">
<h3><span class="header-section-number">4.1.1</span> Reading In Counts</h3>
<p>Now that we’ve defined our set of peaks present under either condition, we can use these to perform a differential signal analysis.
Existing packages such as <code>DiffBind</code> use fixed-width regions to minimise bias from variable sized regions.
Using <code>extraChIPs</code> we can define these ranges using our peak centres that we’ve carried forward.</p>
<pre class="r"><code>centred_peaks &lt;- union_peaks %&gt;% 
  mutate(
    centre = paste0(seqnames, &quot;:&quot;, centre),
    union_peak = granges(.)
  ) %&gt;% 
  colToRanges(var = &quot;centre&quot;) %&gt;% 
  resize(width = 400, fix = &quot;center&quot;) </code></pre>
<p>In the above, we’ve 1) created a new range from the centre-points, 2) placed the original range as a new column in the <code>mcols</code> element, 3) moved the centred range to being the primary range using <code>colToRanges()</code>.
From there, we’ve resized to a standard width of 400bp.
We can now use these ranges to count alignments across our set of samples.
The bam files should be in the <code>data</code> directory and can be formed into a <code>BamFileList</code>, before being passed to <code>regionCounts()</code> from <code>csaw</code>.
This will return a <code>RangedSummarizedExperiment</code> with totals from the entire bam file in the <code>totals</code> column of the <code>colData</code> element</p>
<pre class="r"><code>er_bfl &lt;- file.path(&quot;data&quot;, &quot;ER&quot;, paste0(er_samples$accession, &quot;.bam&quot;)) %&gt;% 
  BamFileList() %&gt;% 
  setNames(er_samples$accession)
er_se &lt;- regionCounts(er_bfl, centred_peaks, ext = 200)
seqinfo(er_se) &lt;- sq
er_se</code></pre>
<pre><code>## class: RangedSummarizedExperiment 
## dim: 188 6 
## metadata(2): final.ext param
## assays(1): counts
## rownames: NULL
## rowData names(7): E2 E2DHT ... gene_name union_peak
## colnames(6): SRR8315180 SRR8315181 ... SRR8315184 SRR8315185
## colData names(4): bam.files totals ext rlen</code></pre>
<pre class="r"><code>colData(er_se)</code></pre>
<pre><code>## DataFrame with 6 rows and 4 columns
##                         bam.files    totals       ext      rlen
##                       &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
## SRR8315180 data/ER/SRR8315180.bam    317845       200        75
## SRR8315181 data/ER/SRR8315181.bam    337623       200        75
## SRR8315182 data/ER/SRR8315182.bam    341998       200        75
## SRR8315183 data/ER/SRR8315183.bam    315872       200        75
## SRR8315184 data/ER/SRR8315184.bam    352908       200        75
## SRR8315185 data/ER/SRR8315185.bam    347709       200        75</code></pre>
<p>Let’s ensure our sample metadata is consistent between this object and <code>er_samples</code></p>
<pre class="r"><code>colData(er_se) &lt;- colData(er_se) %&gt;% 
  as_tibble(rownames = &quot;accession&quot;) %&gt;% 
  left_join(er_samples) %&gt;% 
  as(&quot;DataFrame&quot;)</code></pre>
<p>The use of fixed-width and centred ranges minimises bias which might be between samples with different signal characteristics, but importantly, the counts from these centred ranges as simply our <em>proxy</em> for the ranges defined earlier.
Again, using <code>colToRanges()</code> we can place our original set of union peaks back as the <code>GRanges</code> element underlying the <code>RangedSummarizedExperiment</code>.</p>
<pre class="r"><code>rowRanges(er_se) &lt;- rowRanges(er_se) %&gt;% 
  colToRanges(var = &quot;union_peak&quot;) %&gt;% 
  select(region, starts_with(&quot;gene&quot;))</code></pre>
</div>
<div id="visualising-data" class="section level3" number="4.1.2">
<h3><span class="header-section-number">4.1.2</span> Visualising Data</h3>
<p>Now that we’ve loaded counts, we can explore and visualise them in several ways.
The first might be to check the count distributions, using <code>plotAssayDensities()</code>.</p>
<pre class="r"><code>plotAssayDensities(er_se, trans = &quot;log1p&quot;, colour = &quot;treatment&quot;) +
  scale_colour_manual(values = treat_colours)</code></pre>
<div class="figure"><span style="display:block;" id="fig:plot-er-dentisites"></span>
<img src="index_files/figure-html/plot-er-dentisites-1.png" alt="Densities of raw counts for all centred regions, using the 'log1p' transformation" width="100%" />
<p class="caption">
Figure 6: <span class="caption-title">Densities of raw counts for all centred regions, using the ‘log1p’ transformation</span><br>
</p>
</div>
<p>We might follow this with a PCA</p>
<pre class="r"><code>plotAssayPCA(
  er_se, trans = &quot;log1p&quot;, colour = &quot;treatment&quot;, label = &quot;accession&quot;
) +
  scale_colour_manual(values = treat_colours)</code></pre>
<div class="figure"><span style="display:block;" id="fig:plot-er-pca"></span>
<img src="index_files/figure-html/plot-er-pca-1.png" alt="PCA plot for all ER samples" width="100%" />
<p class="caption">
Figure 7: <span class="caption-title">PCA plot for all ER samples</span><br>
</p>
</div>
<p>An important QC metric as an RLE plot which finds the median value across all samples &amp; ranges to identify any samples with consistently higher/lower or divergent signal.
By way of example, we may wish to normalise out counts using logCPM values for this purpose, so let’s perform this step first</p>
<pre class="r"><code>assay(er_se, &quot;logCPM&quot;) &lt;- cpm(
  assay(er_se, &quot;counts&quot;), lib.size = er_se$totals, log = TRUE
)
plotAssayRle(er_se, assay = &quot;logCPM&quot;, fill = &quot;treatment&quot;) +
  geom_hline(yintercept = 0, linetype = 2) +
  scale_fill_manual(values = treat_colours) </code></pre>
<div class="figure"><span style="display:block;" id="fig:plot-er-rle"></span>
<img src="index_files/figure-html/plot-er-rle-1.png" alt="RLE plot for all ER samples" width="100%" />
<p class="caption">
Figure 8: <span class="caption-title">RLE plot for all ER samples</span><br>
</p>
</div>
<p>A problem sometimes faced in ChIP-Seq data which is less prevalent in RNA-Seq data that counts may be drawn from different distributions under different treatments, as DNA-binding is either increased or decreased globally by a treatment.
We can simply view all samples summarised by treatment group in order to highlight any possible global differences.</p>
<pre class="r"><code>plotAssayRle(er_se, assay = &quot;logCPM&quot;, fill = &quot;treatment&quot;, by_x = &quot;treatment&quot;) +
  geom_hline(yintercept = 0, linetype = 2) +
  scale_fill_manual(values = treat_colours) </code></pre>
<div class="figure"><span style="display:block;" id="fig:plot-er-rle-grouped"></span>
<img src="index_files/figure-html/plot-er-rle-grouped-1.png" alt="RLE plot for ER samples grouped by treatment." width="100%" />
<p class="caption">
Figure 9: <span class="caption-title">RLE plot for ER samples grouped by treatment</span><br>
</p>
</div>
</div>
<div id="differential-signal-analysis" class="section level3" number="4.1.3">
<h3><span class="header-section-number">4.1.3</span> Differential Signal Analysis</h3>
<p>Performing Differential Signal Analysis is now simple using <code>fitAssayDiff()</code>, which offers two methods of analysis: <code>method = "qlf"</code> (default) or <code>method = "lt"</code>.
Respectively, these apply Quasi-Likelihood fits using edgeR <span class="citation">(<a href="#ref-Lund2012-xo">Lund et al. 2012</a>)</span>, as is appropriate for count data, or <code>limma-trend</code> <span class="citation">(<a href="#ref-Law2014-xq">Law et al. 2014</a>)</span> if fitting on the log-transformed, (e.g. <code>logCPM</code>) data.
By default, this will return a <code>SummarizedExperiment</code> with results added to the <code>rowRanged()</code>/<code>rowData()</code> element.
Alternatively, we can return a <code>GRanges</code> object by setting <code>asRanges = TRUE</code></p>
<pre class="r"><code>X &lt;- model.matrix(~treatment, data = colData(er_se))
er_results &lt;- fitAssayDiff(er_se, design = X, asRanges = TRUE)</code></pre>
<p>The default normalisation is to use Library Size as contained in the <code>totals</code> column.
However, by setting <code>lib.size = NULL</code> then <code>colSums()</code> will be used if this is considered more suitable.
Additionally, any of the normalisation methods available in <code>edgeR::calcNormFactors()</code> can be applied.
Providing a column name to the <code>groups</code> argument will also normalise within groups, instead of across the entire dataset.</p>
<pre class="r"><code>er_results &lt;- fitAssayDiff(er_se, design = X, norm = &quot;TMM&quot;, asRanges = TRUE)</code></pre>
<p>Finally, a range-based H<sub>0</sub> <span class="citation">(<a href="#ref-McCarthy2009-qf">McCarthy and Smyth 2009</a>)</span> can be applied by provided a fold-change value below which we consider change to be uninteresting, and we can also automatically add Differential Signal status.</p>
<pre class="r"><code>er_results &lt;- fitAssayDiff(
  er_se, design = X, norm = &quot;TMM&quot;, asRanges = TRUE, fc = 1.2
) %&gt;% 
  addDiffStatus()
arrange(er_results, PValue)</code></pre>
<pre><code>## GRanges object with 188 ranges and 8 metadata columns:
##         seqnames            ranges strand |                   region
##            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |                 &lt;factor&gt;
##     [1]    chr10 81101906-81102928      * | Upstream Promoter (-5kb)
##     [2]    chr10 79629641-79630271      * | Promoter (-2500/+500)   
##     [3]    chr10 89407752-89408138      * | Intron                  
##     [4]    chr10 52233596-52233998      * | Exon                    
##     [5]    chr10 51547205-51547782      * | Promoter (-2500/+500)   
##     ...      ...               ...    ... .                      ...
##   [184]    chr10 93882729-93883168      * |    Intron               
##   [185]    chr10 71344398-71344749      * |    Intergenic (&lt;10kb)   
##   [186]    chr10 76586249-76586887      * |    Promoter (-2500/+500)
##   [187]    chr10 82725993-82726916      * |    Intergenic (&gt;10kb)   
##   [188]    chr10 64605136-64605347      * |    Intron               
##                                            gene_id
##                                    &lt;CharacterList&gt;
##     [1]                       ENSG00000108179.14_6
##     [2]                      ENSG00000151208.17_11
##     [3]   ENSG00000225913.2_9,ENSG00000196566.2_10
##     [4] ENSG00000198964.14_10,ENSG00000225303.2_10
##     [5]                        ENSG00000263639.7_7
##     ...                                        ...
##   [184]                      ENSG00000107864.15_12
##   [185]                        ENSG00000230755.1_6
##   [186]  ENSG00000272748.1_8,ENSG00000156650.14_17
##   [187]                        ENSG00000227209.1_5
##   [188]                        ENSG00000289487.1_1
##                               gene_name       logFC    logCPM      PValue
##                         &lt;CharacterList&gt;   &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;
##     [1]                            PPIF    1.887377   8.02941 6.00421e-11
##     [2]                            DLG5    0.966791   8.23125 4.13714e-06
##     [3] ENSG00000225913,ENSG00000196566    1.712374   6.20441 1.00877e-05
##     [4]           SGMS1,ENSG00000225303    1.129786   6.70351 2.34471e-04
##     [5]                            MSMB    0.664784   7.44213 1.13493e-03
##     ...                             ...         ...       ...         ...
##   [184]                           CPEB3 -0.00513719   6.39992    0.985018
##   [185]                       MTATP6P23 -0.00519869   6.77464    0.985285
##   [186]           ENSG00000272748,KAT6B -0.00909379   8.12097    0.985618
##   [187]                         WARS2P1 -0.00371918   9.43978    0.994243
##   [188]                 ENSG00000289487  0.00109980   5.96820    0.997631
##                 FDR    status
##           &lt;numeric&gt;  &lt;factor&gt;
##     [1] 1.12879e-08 Increased
##     [2] 3.88891e-04 Increased
##     [3] 6.32163e-04 Increased
##     [4] 1.10201e-02 Increased
##     [5] 4.26732e-02 Increased
##     ...         ...       ...
##   [184]    0.996217 Unchanged
##   [185]    0.996217 Unchanged
##   [186]    0.996217 Unchanged
##   [187]    0.997631 Unchanged
##   [188]    0.997631 Unchanged
##   -------
##   seqinfo: 1 sequence from hg19 genome</code></pre>
<p>Given these results are in a <code>GRanges</code> object, we can produce an MA-plot using conventional <code>ggplot2</code> approaches.</p>
<pre class="r"><code>status_colours &lt;- c(
  Undetected = &quot;grey70&quot;, Unchanged = &quot;grey30&quot;, 
  Decreased = &quot;blue&quot;, Increased = &quot;red&quot;
)
er_results %&gt;% 
  as_tibble() %&gt;% 
  ggplot(aes(logCPM, logFC)) +
  geom_point(aes(colour = status)) +
  geom_smooth(method = &quot;lm&quot;, se = FALSE) +
  scale_colour_manual(values = status_colours)</code></pre>
<div class="figure"><span style="display:block;" id="fig:er-plotma"></span>
<img src="index_files/figure-html/er-plotma-1.png" alt="MA-plot for ERa analysis using fixed-width windows, TMM normalisation and setting 20% as the range beyond which we consider diferential signal to be of interest." width="100%" />
<p class="caption">
Figure 10: <span class="caption-title">MA-plot for ERa analysis using fixed-width windows, TMM normalisation and setting 20% as the range beyond which we consider diferential signal to be of interest</span><br>
</p>
</div>
</div>
<div id="visualising-results" class="section level3" number="4.1.4">
<h3><span class="header-section-number">4.1.4</span> Visualising Results</h3>
<p>Additionally, we might like to compare Differential Signal Status against the regions we defined earlier and
<code>plotSplitDonut()</code> is well suited to this.
A key advantage of <code>plotSplitDonut()</code> is the ability to specify inner and outer palettes separately</p>
<pre class="r"><code>er_results %&gt;% 
  plotSplitDonut(
    inner = &quot;region&quot;, outer = &quot;status&quot;,
    inner_palette = region_colours, outer_palette = status_colours
  )</code></pre>
<div class="figure"><span style="display:block;" id="fig:plotsplitdonut-default"></span>
<img src="index_files/figure-html/plotsplitdonut-default-1.png" alt="Split-Donut plot showing distribution of peaks with diferential signal" width="100%" />
<p class="caption">
Figure 11: <span class="caption-title">Split-Donut plot showing distribution of peaks with diferential signal</span><br>
</p>
</div>
<p>Once again, labels are controlled using the <code>glue</code> syntax and segments are able to be exploded based on matches to a regular expression</p>
<pre class="r"><code>er_results %&gt;% 
  plotSplitDonut(
    inner = &quot;region&quot;, outer = &quot;status&quot;,
    inner_palette = region_colours, outer_palette = status_colours,
    inner_glue = &quot;n = {n}&quot;, inner_min_p = 0, inner_label_size = 3.5,
    outer_glue = &quot;{str_wrap(.data[[inner]], 15)}\n{.data[[outer]]}\nn = {n}&quot;,
    outer_min_p = 0, outer_max_p = 0.03,
    explode_outer = &quot;Increased|Decreased&quot;, explode_r = 0.3
  )</code></pre>
<div class="figure"><span style="display:block;" id="fig:plotsplitdonut-labels"></span>
<img src="index_files/figure-html/plotsplitdonut-labels-1.png" alt="Split-Donut plot exploding segments with evidence of differential signal, and modifying segment labels" width="100%" />
<p class="caption">
Figure 12: <span class="caption-title">Split-Donut plot exploding segments with evidence of differential signal, and modifying segment labels</span><br>
</p>
</div>
</div>
</div>
<div id="using-sliding-windows" class="section level2" number="4.2">
<h2><span class="header-section-number">4.2</span> Using Sliding Windows</h2>
<p><code>fitAssayDiff()</code> is agnostic to sliding or fixed-width windows.
Whilst fixed-width windows are relatively simple conceptually, sliding windows offer other advantages such as being able to manage data where signal comes from genomic regions which are highly variable in width, such as histone marks like H3K27ac.
Sliding window analysis is well-implemented in <code>csaw</code>, however <code>extraChIPs</code> offers the additional ability to combine ranges using asymptotically correct <em>harmonic mean p-value</em> <span class="citation">(<a href="#ref-Wilson2019-ln">Wilson 2019</a>)</span>.
The step of moving from genomic windows to a subset of windows which we can confidently assume contain signal is also able to be managed by <code>extraChIPs</code> using <code>dualFilter()</code>.</p>
<p>For this section of the analysis, we’ll use data for the histone mark H3K27ac from the same cell type under the same two conditions (E2 + E2DHT).
A new sample metadata sheet can be formed using the following, and we’ll immediately for the <code>BamFileList</code> for reading in counts.</p>
<pre class="r"><code>h3k_samples &lt;- tibble(
  accession = paste0(&quot;SRR83151&quot;, seq(86, 91)),
  target = &quot;H3K27ac&quot;,
  treatment = c(factor(rep(treat_levels, each = 3), levels = treat_levels))
) %&gt;% 
  bind_rows(
    tibble(accession = &quot;SRR8315192&quot;, target = &quot;Input&quot;)
  )
h3k_bfl &lt;- file.path(
  &quot;data&quot;, &quot;H3K27ac&quot;, paste0(h3k_samples$accession, &quot;.bam&quot;)
) %&gt;% 
  BamFileList() %&gt;% 
  setNames(h3k_samples$accession)</code></pre>
<div id="defining-windows" class="section level3" number="4.2.1">
<h3><span class="header-section-number">4.2.1</span> Defining Windows</h3>
<p>Given we’re going to use sliding windows, we can use a smaller window size and need to set a step-size for ‘sliding’ along the genome.
Whilst there are no hard and fast rules, for a histone marks, window sizes of 150bp sliding in steps of 50bp may be a good choice.</p>
<pre class="r"><code>win_size &lt;- 150
win_step &lt;- 50</code></pre>
<p>The example dataset for today can be easily managed on a laptop, however, it should be noted that for a full-scale dataset, large amounts of RAM are required, and this process may take considerable time.
In the following we define some key parameters for passing to <code>csaw::windowCounts()</code>, where we restrict counting to <code>chr10</code>, provide out black and grey-lists and let the function know that these are all single-end reads.</p>
<p>From there, we perform the counts using the window step &amp; size as defined above, extending each read to represent a full DNA fragment of 200nt, and discarding any regions with fewer than 1 read/sample.</p>
<pre class="r"><code>rp &lt;- readParam(pe = &quot;none&quot;, restrict = &quot;chr10&quot;, discard = bg_list)
wincounts &lt;- windowCounts(
  bam.files = h3k_bfl,
  spacing = win_step, width = win_size, ext = 200,
  filter = length(h3k_bfl),
  param = rp
)
seqlevels(wincounts) &lt;- seqlevels(sq)
seqinfo(wincounts) &lt;- sq
dim(wincounts)</code></pre>
<pre><code>## [1] 467904      7</code></pre>
<p>Many of these windows will have counted reads which are essentially background noise, and we wish to exclude these and only retain those which confidently contain signal from our ChIP target.
<code>extraChIPs</code> provides a function <code>dualFilter()</code> in which a set of guide ranges can be provided where signal is expected.
Thresholds for minimum overall signal, and signal relative to input are set based on retaining a give proportion of reads which overlap these ranges, with lower numbers returning more stringently selected ranges.</p>
<p>A set of guide ranges has been prepared following a similar process to forming union peaks above for ER<span class="math inline">\(\alpha\)</span>, and can be loaded then passed to <code>dualFilter()</code>.
Again, this dataset is easily handled on a laptop, whilst for a complete genome, considerable amounts of RAM will be required.</p>
<pre class="r"><code>guide_ranges &lt;- file.path(&quot;data&quot;, &quot;H3K27ac&quot;, &quot;H3K27ac_chr10.bed&quot;) %&gt;% 
  import.bed(seqinfo = sq)
filtcounts &lt;- dualFilter(
  x = wincounts[,dplyr::filter(h3k_samples, target == &quot;H3K27ac&quot;)$accession],
  bg  =  wincounts[,dplyr::filter(h3k_samples, target == &quot;Input&quot;)$accession],
  ref = guide_ranges, q = 0.6
)
colData(filtcounts) &lt;- colData(filtcounts) %&gt;% 
  as_tibble(rownames = &quot;accession&quot;) %&gt;% 
  left_join(h3k_samples) %&gt;% 
  as(&quot;DataFrame&quot;)
dim(filtcounts)</code></pre>
<pre><code>## [1] 19017     6</code></pre>
<p>By default, an additional assay <code>logCPM</code> will be added during this step and this can also be used for QC.</p>
<pre class="r"><code>A &lt;- plotAssayDensities(filtcounts, assay = &quot;logCPM&quot;, colour = &quot;treatment&quot;) + 
  scale_colour_manual(values = treat_colours)
B &lt;- plotAssayPCA(
  filtcounts, assay = &quot;logCPM&quot;, colour = &quot;treatment&quot;, label = &quot;accession&quot;
) +
  scale_colour_manual(values = treat_colours)
A + B + plot_layout(guides = &quot;collect&quot;) + plot_annotation(tag_levels = &quot;A&quot;) </code></pre>
<div class="figure"><span style="display:block;" id="fig:plot-filt-assays"></span>
<img src="index_files/figure-html/plot-filt-assays-1.png" alt="A) Density plot and B) PCA based on logCPM values from the sliding windows retained after filtering." width="100%" />
<p class="caption">
Figure 13: <span class="caption-title">A) Density plot and B) PCA based on logCPM values from the sliding windows retained after filtering</span><br>
</p>
</div>
<p>Now we can pass our filtered windows to <code>fitAssayDiff()</code> as we did previously, however, once we have performed statistical testing on each window, we need to merge windows to obtain our final results.</p>
<pre class="r"><code>X &lt;- model.matrix(~treatment, data = colData(filtcounts))
h3k_all &lt;- fitAssayDiff(filtcounts, norm = &quot;TMM&quot;, design = X, asRanges = TRUE)</code></pre>
<p>Although we haven’t declared in regions as showing differential signal, we can still inspect an MA plot for any unexpected issues in our data</p>
<pre class="r"><code>h3k_all %&gt;% 
  as_tibble() %&gt;% 
  ggplot(aes(logCPM, logFC)) +
  geom_point() +
  geom_smooth(se = FALSE)</code></pre>
<div class="figure"><span style="display:block;" id="fig:plotma-h3k"></span>
<img src="index_files/figure-html/plotma-h3k-1.png" alt="MA plot using all sliding windows retained after the initial filtering step." width="100%" />
<p class="caption">
Figure 14: <span class="caption-title">MA plot using all sliding windows retained after the initial filtering step</span><br>
</p>
</div>
<p>As expected with H3K27ac signal, the total amount of signal obtained under two related conditions is similar and we have a nicely centred plot.</p>
</div>
<div id="merging-windows" class="section level3" number="4.2.2">
<h3><span class="header-section-number">4.2.2</span> Merging Windows</h3>
<p>Now we can merge overlapping windows to obtain our final results.
Using <code>extraChIPs</code> this can be performed using all methods offered by <code>csaw</code>, (<code>?mergeByCol</code> and <code>?mergeBySig</code>) as well as using the asymptotically-exact harmonic mean p-value, which we’ll use here.</p>
<pre class="r"><code>h3k_results &lt;- mergeByHMP(h3k_all, min_win = 2, merge_within = win_size + 1) %&gt;% 
  addDiffStatus()
h3k_results</code></pre>
<pre><code>## GRanges object with 643 ranges and 9 metadata columns:
##         seqnames            ranges strand | n_windows      n_up    n_down
##            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##     [1]    chr10 42862951-42863350      * |         6         5         0
##     [2]    chr10 43047801-43048200      * |         6         0         3
##     [3]    chr10 43048401-43048800      * |         6         0         3
##     [4]    chr10 43132901-43134800      * |        36        13        16
##     [5]    chr10 43135701-43137250      * |        29         7         3
##     ...      ...               ...    ... .       ...       ...       ...
##   [639]    chr10 99473501-99474250      * |        13         0         5
##   [640]    chr10 99496401-99497850      * |        27         9        18
##   [641]    chr10 99608301-99610050      * |        32         0         8
##   [642]    chr10 99628751-99629400      * |        11         2         6
##   [643]    chr10 99894101-99895400      * |        24         4        13
##                    keyval_range    logCPM      logFC       hmp   hmp_fdr
##                       &lt;GRanges&gt; &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
##     [1] chr10:42863051-42863200   6.31662  0.3345653 0.4041174  0.997968
##     [2] chr10:43048001-43048150   6.16996 -0.0780895 0.8764236  0.997968
##     [3] chr10:43048601-43048750   6.31199 -0.2619553 0.6083939  0.997968
##     [4] chr10:43134251-43134400   7.67436  0.1350775 0.5145484  0.997968
##     [5] chr10:43137101-43137250   6.87832  0.0791231 0.0119809  0.126290
##     ...                     ...       ...        ...       ...       ...
##   [639] chr10:99473851-99474000   6.43974 -0.4427309 0.2576771  0.815392
##   [640] chr10:99497701-99497850   7.41777  0.0551099 0.9923554  0.997968
##   [641] chr10:99608601-99608750   7.10450 -0.5008959 0.0309346  0.236797
##   [642] chr10:99628851-99629000   6.63987 -0.2298606 0.6588732  0.997968
##   [643] chr10:99895051-99895200   8.08365 -0.1143540 0.6362419  0.997968
##            status
##          &lt;factor&gt;
##     [1] Unchanged
##     [2] Unchanged
##     [3] Unchanged
##     [4] Unchanged
##     [5] Unchanged
##     ...       ...
##   [639] Unchanged
##   [640] Unchanged
##   [641] Unchanged
##   [642] Unchanged
##   [643] Unchanged
##   -------
##   seqinfo: 1 sequence from hg19 genome</code></pre>
<p>Notice that we have additional columns not returned when using fixed-width windows.
The first three indicate how many original windows were merged, along with how many may have been individually considered as showing differential signal in either direction.
Next to these is the column <code>keyval_range</code> which indicates the original window which returned the lowest p-value, or by setting <code>keyval = "merged"</code>, will return the range containing all original windows with raw p-values below the harmonic-mean p-value.</p>
<p>Given the harmonic mean involves summing the inverse of raw p-values, these are considered to be weights and the returned values for logCPM and logFC are also weighted averages, using <span class="math inline">\(w_i = \frac{1}{p_i}\)</span> as weights for each windows <span class="math inline">\(i\)</span>.
Given this, any MA-plots may appear to show bias, which is in fact due to the weighting strategy used here.
As such the previous figure using all windows prior to merging is the best strategy for identifying any issues with the data</p>
<pre class="r"><code>h3k_results %&gt;% 
  as_tibble() %&gt;% 
  ggplot(aes(logCPM, logFC)) +
  geom_point(aes(colour = status), alpha = 0.7) +
  geom_smooth(method = &quot;lm&quot;, se = FALSE) +
  scale_colour_manual(values = status_colours)</code></pre>
<div class="figure"><span style="display:block;" id="fig:plotma-merged-h3k"></span>
<img src="index_files/figure-html/plotma-merged-h3k-1.png" alt="MA plot showing results after merging overlapping windows." width="100%" />
<p class="caption">
Figure 15: <span class="caption-title">MA plot showing results after merging overlapping windows</span><br>
</p>
</div>
</div>
</div>
</div>
<div id="comparing-both-sets-of-results" class="section level1" number="5">
<h1><span class="header-section-number">5</span> Comparing Both Sets Of Results</h1>
<p><code>extraChIPs</code> provide additional capacity for comparing between two or more sets of results.
The key structure for this is a <code>GRangesList</code>, so let’s ensure we have compatible column names, then form a combined object.</p>
<pre class="r"><code>h3k_results &lt;- h3k_results %&gt;% 
  select(logCPM, logFC, PValue = hmp, FDR = hmp_fdr, status) %&gt;% 
  mutate(region = bestOverlap(., unlist(regions), var = &quot;region&quot;)) %&gt;% 
  mapByFeature(genes = gtf$gene, prom = regions$promoter)
both_results &lt;- GRangesList(ER = er_results, H3K27ac = h3k_results) </code></pre>
<div id="association-between-datasets" class="section level2" number="5.1">
<h2><span class="header-section-number">5.1</span> Association Between Datasets</h2>
<p>Now we have a combined <code>GRangesList</code> object, we can plot columns from each element against the other.
If we wish to compare signal in sites where both are detected, we can use <code>plotPairwise()</code>, which will also show boxplots of signal for where only one target is detected, alongside those where both targets are detected</p>
<pre class="r"><code>plotPairwise(both_results, var = &quot;logCPM&quot;)</code></pre>
<div class="figure"><span style="display:block;" id="fig:plot-pairwise-logcpm"></span>
<img src="index_files/figure-html/plot-pairwise-logcpm-1.png" alt="Comparison of logCPM values when both targets are detected, with boxplots showing the range of values when each target was detected alone, or when both were detected together" width="100%" />
<p class="caption">
Figure 16: <span class="caption-title">Comparison of logCPM values when both targets are detected, with boxplots showing the range of values when each target was detected alone, or when both were detected together</span><br>
</p>
</div>
<p>Although we only have a handful of sites in our example dataset, one might easily suspect that signal for both targets is higher when both are detected together.</p>
<p>Alternatively, we might wish to see which sites are changing together, or where only one is changing.
This time we’ll plot <code>logFC</code>, colouring points by the <em>combined</em> Differential Signal status in each element, and adding labels from the <code>gene_name</code> column.
Labels are only added to the sites in each combined group which are furthest from zero.</p>
<pre class="r"><code>plotPairwise(
  both_results, var = &quot;logFC&quot;, colour = &quot;status&quot;, label = &quot;gene_name&quot;
) +
  scale_colour_brewer(palette = &quot;Set1&quot;, direction = -1)</code></pre>
<div class="figure"><span style="display:block;" id="fig:plot-pairwise-logFC"></span>
<img src="index_files/figure-html/plot-pairwise-logFC-1.png" alt="Comparison of logFC values when both targets are detcted, with boxplots showing the range of values when only one target is detected, or when both are detected" width="100%" />
<p class="caption">
Figure 17: <span class="caption-title">Comparison of logFC values when both targets are detcted, with boxplots showing the range of values when only one target is detected, or when both are detected</span><br>
</p>
</div>
<p>We can also compare columns by performing an operation somewhat analogous to <code>pivot_wider</code> where we find overlapping ranges and add columns from each element to a set of consensus peaks.</p>
<pre class="r"><code>mapGrlCols(both_results, var = &quot;status&quot;) </code></pre>
<pre><code>## GRanges object with 741 ranges and 2 metadata columns:
##         seqnames            ranges strand | ER_status H3K27ac_status
##            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;factor&gt;       &lt;factor&gt;
##     [1]    chr10 42862951-42863350      * | NA             Unchanged
##     [2]    chr10 43047801-43048800      * | Unchanged      Unchanged
##     [3]    chr10 43047801-43048800      * | Unchanged      Unchanged
##     [4]    chr10 43132901-43134800      * | NA             Unchanged
##     [5]    chr10 43135701-43137250      * | NA             Unchanged
##     ...      ...               ...    ... .       ...            ...
##   [737]    chr10 99496401-99497850      * | NA             Unchanged
##   [738]    chr10 99608301-99610050      * | NA             Unchanged
##   [739]    chr10 99621632-99621988      * | Unchanged      NA       
##   [740]    chr10 99628751-99629400      * | NA             Unchanged
##   [741]    chr10 99894101-99895400      * | NA             Unchanged
##   -------
##   seqinfo: 1 sequence from hg19 genome</code></pre>
<p>We can use this to quickly find sites which are of interest in both sets of results, along with their targets</p>
<pre class="r"><code>both_results %&gt;% 
  mapGrlCols(var = c(&quot;region&quot;, &quot;status&quot;, &quot;FDR&quot;), collapse = &quot;gene_name&quot;) %&gt;% 
  filter(
    str_detect(ER_status, &quot;Increased|Decreased&quot;),
    str_detect(H3K27ac_status, &quot;Increased|Decreased&quot;)
  )</code></pre>
<pre><code>## GRanges object with 3 ranges and 7 metadata columns:
##       seqnames            ranges strand |                ER_region
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |                 &lt;factor&gt;
##   [1]    chr10 71879101-71881150      * | Promoter (-2500/+500)   
##   [2]    chr10 79623301-79635950      * | Promoter (-2500/+500)   
##   [3]    chr10 81101651-81103150      * | Upstream Promoter (-5kb)
##                 H3K27ac_region ER_status H3K27ac_status      ER_FDR H3K27ac_FDR
##                       &lt;factor&gt;  &lt;factor&gt;       &lt;factor&gt;   &lt;numeric&gt;   &lt;numeric&gt;
##   [1] Promoter (-2500/+500)    Increased      Increased 4.46675e-02 8.68739e-03
##   [2] Promoter (-2500/+500)    Increased      Increased 3.88891e-04 7.76644e-06
##   [3] Upstream Promoter (-5kb) Increased      Increased 1.12879e-08 1.60656e-05
##                   gene_name
##                 &lt;character&gt;
##   [1]                 AIFM2
##   [2] DLG5; ENSG00000204049
##   [3]                  PPIF
##   -------
##   seqinfo: 1 sequence from hg19 genome</code></pre>
<p>This also makes comparison of sites easy using <code>plotSplitDonut()</code> and in the following we’ll only show sites where both are detected.</p>
<pre class="r"><code>both_results %&gt;% 
  mapGrlCols(var = &quot;status&quot;) %&gt;% 
  as_tibble() %&gt;% 
  dplyr::filter(!if_any(ends_with(&quot;status&quot;), is.na)) %&gt;% 
  dplyr::rename_with(\(x) str_remove_all(x, &quot;_status&quot;)) %&gt;% 
  plotSplitDonut(
    inner = &quot;H3K27ac&quot;, outer = &quot;ER&quot;,
    inner_palette = status_colours, outer_palette = status_colours,
    inner_glue = &quot;{inner}\n{.data[[inner]]}\n{n}&quot;,
    outer_glue = &quot;{outer}\n{.data[[outer]]}\n{n}&quot;,
    label_alpha = 0.8, min_p = 0.02,
    explode_inner = &quot;Increased&quot;, explode_outer = &quot;Increased&quot;, 
    explode_query = &quot;AND&quot;, explode_r = 0.3
  )</code></pre>
<div class="figure"><span style="display:block;" id="fig:plot-status-donut"></span>
<img src="index_files/figure-html/plot-status-donut-1.png" alt="Combined status when both targets are detected together" width="100%" />
<p class="caption">
Figure 18: <span class="caption-title">Combined status when both targets are detected together</span><br>
</p>
</div>
<p>Perhaps we might wish to look at the combined status in reference to the genomic regions we defined previously.
Here, we’ll scale the site by width to show results in kb, instead of counting sites.</p>
<pre class="r"><code>both_results %&gt;% 
  mapGrlCols(var = c(&quot;region&quot;, &quot;status&quot;)) %&gt;% 
  as_tibble(rangeAsChar = FALSE) %&gt;% 
  dplyr::filter(!if_any(ends_with(&quot;status&quot;), is.na)) %&gt;% 
  mutate(
    ER_status = fct_relabel(ER_status, \(x) paste(&quot;ER&quot;, x)),
    H3K27ac_status = fct_relabel(H3K27ac_status, \(x) paste(&quot;H3K27ac&quot;, x)),
    combined = fct_cross(ER_status, H3K27ac_status, sep = &quot;\n&quot;)
  ) %&gt;% 
  plotSplitDonut(
    scale_by = &quot;width&quot;,
    inner = &quot;H3K27ac_region&quot;, outer = &quot;combined&quot;,
    inner_palette = region_colours,
    total_glue = &quot;{comma(N)}kb&quot;,
    inner_glue = &quot;{str_wrap(.data[[inner]], 15)}\n{comma(n, 1)}kb&quot;,
    outer_glue = &quot;{.data[[outer]]}\n{comma(n, 0.1)}kb&quot;,
    outer_min_p = 0.01, #outer_max_p = 0.02,
    explode_outer = &quot;(Inc|Dec).+(Inc|Dec)&quot;, explode_r = 0.4,
    label_alpha = 0.7
  )</code></pre>
<div class="figure"><span style="display:block;" id="fig:plot-stauts-region-donut"></span>
<img src="index_files/figure-html/plot-stauts-region-donut-1.png" alt="Combined status shown by the region, as defined by the best overlap for H3K27ac signal. Regions here are scaled by width and totals shown in kb" width="100%" />
<p class="caption">
Figure 19: <span class="caption-title">Combined status shown by the region, as defined by the best overlap for H3K27ac signal</span><br>Regions here are scaled by width and totals shown in kb
</p>
</div>
</div>
<div id="regions-in-context" class="section level2" number="5.2">
<h2><span class="header-section-number">5.2</span> Regions in Context</h2>
<p>As well the above strategies for inspecting results, looking directly at the sites using a genome browser is a common strategy.
<code>extraChIPs</code> provides the function <code>plotHFGC()</code> which enables us to take a genome level viewpoint of any 1) HiC Interactions, 2) Features, 3) Gene Models and 4) Coverage.
All tracks are optional and will be displayed in this order when provided using the various track types provided by <code>Gviz</code> <span class="citation">(<a href="#ref-gviz">Hahne and Ivanek 2016</a>)</span>.
In order to use this approach to scan through our results choosing one ranges at a time, let’s define the elements that we have starting with our genomic regions, which we would consider to be a set of genomic features.
For this track, we’ll need a <code>GRangesList()</code> and each element will be filled a different colour, meaning we’ll need a matching vector of colours.
We already have both of these from earlier on our workflow, but we will need to change the names on our colour vector</p>
<pre class="r"><code>names(region_colours) &lt;- names(regions)</code></pre>
<p>Next we’ll setup our gene models in the format that <code>Gviz</code> is familiar with, using the <code>exons</code> element from our GTF annotations.</p>
<pre class="r"><code>gene_models &lt;- gtf$exon %&gt;% 
  select(
    type, gene = gene_id, exon = exon_id, transcript = transcript_id, 
    symbol = gene_name
  )</code></pre>
<p>This leaves coverage as our final track, or set of tracks.
<code>plotHFGC()</code> takes <code>BigWigFileList</code> objects for coverage and if passing a single <code>BigwigFileList</code>, each element will be drawn on a separate track.
For our results we have both ER and H3K27ac so a more informative plot might be to overlay the two treatments in a single track for each target.
To facilitate this, we need a <em>named</em> list of <code>BigwigFileList</code> objects and a matching <em>named</em> list of colours.
In each of our data directories we have summarised coverage for E2 and E2DHT</p>
<pre class="r"><code>targets &lt;- c(&quot;ER&quot;, &quot;H3K27ac&quot;)
bwfl &lt;- list(
  ER = file.path(&quot;data&quot;, &quot;ER&quot;, glue(&quot;{treat_levels}_cov_chr10.bw&quot;)),
  H3K27ac = file.path(&quot;data&quot;, &quot;H3K27ac&quot;, glue(&quot;{treat_levels}_cov_chr10.bw&quot;))
) %&gt;% 
  lapply(setNames, treat_levels) %&gt;% 
  lapply(BigWigFileList)
line_colours &lt;- lapply(bwfl, function(x) treat_colours[names(x)])</code></pre>
<p>Now we have our track setup, we’ll load the cytogenetic bands provided with <code>extraChIPs</code> to ensure we have a band at the top of the plot, although this too is optional.</p>
<pre class="r"><code>data(&quot;grch37.cytobands&quot;)</code></pre>
<p>And finally, let’s select our first range as the most highly ranked site for change ER binding, which also appears to show an increase in H3K27ac signal</p>
<pre class="r"><code>gr &lt;- both_results %&gt;% 
  mapGrlCols(
    var = c(&quot;region&quot;, &quot;status&quot;, &quot;FDR&quot;), collapse = &quot;gene_name&quot;
  ) %&gt;% 
  arrange(ER_FDR) %&gt;% 
  .[1]</code></pre>
<p>Now we pass all of these to <code>plotHFGC()</code> zooming out a little to try and get some kind of wider context.</p>
<pre class="r"><code>plotHFGC(
  gr, cytobands = grch37.cytobands, 
  features = regions, featcol = region_colours,
  genes = gene_models, genecol = &quot;wheat&quot;, 
  coverage =  bwfl, linecol = line_colours,
  zoom = 10
)</code></pre>
<div class="figure"><span style="display:block;" id="fig:plot-hfgc-default"></span>
<img src="index_files/figure-html/plot-hfgc-default-1.png" alt="Default plot from plotHFGC showing both ER and H3K27ac signal on separate tracks, with the two treatment groups overlaid as separately coloured lines." width="100%" />
<p class="caption">
Figure 20: <span class="caption-title">Default plot from plotHFGC showing both ER and H3K27ac signal on separate tracks, with the two treatment groups overlaid as separately coloured lines</span><br>
</p>
</div>
<p>One helpful feature of is the ability to annotate our coverage tracks to indicate statistical significance.
If we simply split our initial set of results into <code>GRangesList</code> objects based on their status, we can use these to add coloured bars above our coverage tracks to denote that signal was detected there, and what the results from analysis were.
We can simply pass our vector of status colours that we’ve already defined for these</p>
<pre class="r"><code>cov_annot &lt;- list(
  ER = splitAsList(er_results, er_results$status),
  H3K27ac = splitAsList(h3k_results, h3k_results$status)
)</code></pre>
<p>Now let’s add these annotations and tweak the appearance of the plot</p>
<pre class="r"><code>plotHFGC(
  gr, cytobands = grch37.cytobands, 
  features = regions, featcol = region_colours, featstack = &quot;dense&quot;,
  genes = gene_models, genecol = &quot;wheat&quot;,
  coverage =  bwfl, linecol = line_colours,
  annotation = cov_annot, annotcol = status_colours,
  zoom = 10, featsize = 2, title.width = 0.9, 
  cex.axis = 1, cex.title = 1, rotation.title = 90,  fontsize = 12
)</code></pre>
<div class="figure"><span style="display:block;" id="fig:plot-hfgc-final"></span>
<img src="index_files/figure-html/plot-hfgc-final-1.png" alt="Customised output from plotHFGC, adding annotations for each site as well as tweaking the labels and font sizes" width="100%" />
<p class="caption">
Figure 21: <span class="caption-title">Customised output from plotHFGC, adding annotations for each site as well as tweaking the labels and font sizes</span><br>
</p>
</div>
<p>Just like we were able to provide a separate <code>BigWigFileList</code> to indicate separate tracks for our targets, we can use a similar strategy for the genes and features tracks.</p>
<ol style="list-style-type: decimal">
<li>A <code>GRangesList</code> of gene will draw each element on a separate track. This may be useful for indicating results from a Differential Gene Expression analysis by plotting up-regulated and down-regulated genes on separate tracks in different colours</li>
<li>Passing a <em>named</em> list of <code>GRangesList</code> objects to the features track will add each element of the list as a separate tracks. In this way you may wish to add multiple features, such as a <code>chromHMM</code> track along with CTCF or other known binding sites.</li>
</ol>
</div>
<div id="profile-heatmaps" class="section level2" number="5.3">
<h2><span class="header-section-number">5.3</span> Profile Heatmaps</h2>
<p>A final way of visualising results may be to draw a Profile Heatmap, where we show coverage or fold-enrichment over background in the surrounding regions.
Instead of separating our <code>BigWigFileList</code> into separate tracks, let’s just use a single <code>BigWigFileList</code> with coverage for both ER and H3K27ac.</p>
<pre class="r"><code>profile_bwfl &lt;- bwfl %&gt;% 
  lapply(path) %&gt;% 
  unlist() %&gt;% 
  setNames(str_replace_all(names(.), &quot;\\.&quot;, &quot; &quot;)) %&gt;% 
  BigWigFileList()</code></pre>
<p>Given we only have a small dataset, let’s just choose the top 20 sites by logFC in the ER dataset, and draw Profile Heatmaps for both targets.
We first make a call to <code>getProfileData()</code> which smooths the data into bins around the site</p>
<pre class="r"><code>er_sig &lt;- arrange(er_results, -logFC)[1:20]
pd &lt;- getProfileData(profile_bwfl, er_sig)</code></pre>
<p>We can now pass this to <code>plotProfileHeatmap()</code> and can use any of the categorical columns in the original set of ranges to facet our results.
Here we’ll facet by status in the ER dataset, which will also draw lines in separate colours.</p>
<pre class="r"><code>plotProfileHeatmap(pd, &quot;profile_data&quot;, facetY = &quot;status&quot;) +
  scale_colour_manual(values = status_colours) +
  scale_fill_gradient(low = &quot;white&quot;, high = &quot;red&quot;) +
  labs(fill = &quot;logCPM&quot;)</code></pre>
<div class="figure"><span style="display:block;" id="fig:plot-profile-heatmaps"></span>
<img src="index_files/figure-html/plot-profile-heatmaps-1.png" alt="Profile Heatmap for the 20 most highly ranked sites for increased ER binding. Ranges are centred around the ER binding patterns. The double peak seen earlier in the output from plotHFGC is at the top of the first panel.`" width="100%" />
<p class="caption">
Figure 22: <span class="caption-title">Profile Heatmap for the 20 most highly ranked sites for increased ER binding</span><br>Ranges are centred around the ER binding patterns. The double peak seen earlier in the output from plotHFGC is at the top of the first panel.`
</p>
</div>
<p>A complete dataset would provide a far cleaner set of results, but we can still see the general increases in ER signal and how these relate to H3K27ac signal.</p>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1>References</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-edgeRQLF2016" class="csl-entry">
Chen, Yunshun, Aaron A T Lun, and Gordon K Smyth. 2016. <span>“From Reads to Genes to Pathways: Differential Expression Analysis of RNA-Seq Experiments Using Rsubread and the edgeR Quasi-Likelihood Pipeline.”</span> <em>F1000Research</em> 5: 1438. <a href="https://doi.org/10.12688/f1000research.8987.2">https://doi.org/10.12688/f1000research.8987.2</a>.
</div>
<div id="ref-gviz" class="csl-entry">
Hahne, Florian, and Robert Ivanek. 2016. <span>“Statistical Genomics: Methods and Protocols.”</span> In, edited by Ewy Mathé and Sean Davis, 335–51. New York, NY: Springer New York. <a href="https://doi.org/10.1007/978-1-4939-3578-9_16">https://doi.org/10.1007/978-1-4939-3578-9_16</a>.
</div>
<div id="ref-Hickey2021-mz" class="csl-entry">
Hickey, Theresa E, Luke A Selth, Kee Ming Chia, Geraldine Laven-Law, Heloisa H Milioli, Daniel Roden, Shalini Jindal, et al. 2021. <span>“The Androgen Receptor Is a Tumor Suppressor in Estrogen Receptor-Positive Breast Cancer.”</span> <em>Nat. Med.</em> 27 (2): 310–20.
</div>
<div id="ref-ComplexUpset2020" class="csl-entry">
Krassowski, Michał. 2020. <span>“ComplexUpset.”</span> <a href="https://doi.org/10.5281/zenodo.3700590">https://doi.org/10.5281/zenodo.3700590</a>.
</div>
<div id="ref-Law2014-xq" class="csl-entry">
Law, Charity W, Yunshun Chen, Wei Shi, and Gordon K Smyth. 2014. <span>“Voom: Precision Weights Unlock Linear Model Analysis Tools for <span class="nocase">RNA-seq</span> Read Counts.”</span> <em>Genome Biol.</em> 15 (2): R29.
</div>
<div id="ref-csaw2016" class="csl-entry">
Lun, Aaron T L, and Gordon K Smyth. 2016. <span>“Csaw: A <span>B</span>ioconductor Package for Differential Binding Analysis of <span>C</span>h<span>I</span><span>P</span>-Seq Data Using Sliding Windows.”</span> <em>Nucleic Acids Res.</em> 44 (5): e45.
</div>
<div id="ref-Lund2012-xo" class="csl-entry">
Lund, Steven P, Dan Nettleton, Davis J McCarthy, and Gordon K Smyth. 2012. <span>“Detecting Differential Expression in <span class="nocase">RNA-sequence</span> Data Using Quasi-Likelihood with Shrunken Dispersion Estimates.”</span> <em>Stat. Appl. Genet. Mol. Biol.</em> 11 (5).
</div>
<div id="ref-McCarthy2009-qf" class="csl-entry">
McCarthy, Davis J, and Gordon K Smyth. 2009. <span>“Testing Significance Relative to a Fold-Change Threshold Is a <span>TREAT</span>.”</span> <em>Bioinformatics</em> 25 (6): 765–71.
</div>
<div id="ref-pederson_2023" class="csl-entry">
Pederson, Stevie. 2023. <span>“prepareChIPs:”</span> <a href="https://doi.org/10.48546/WORKFLOWHUB.WORKFLOW.528.1">https://doi.org/10.48546/WORKFLOWHUB.WORKFLOW.528.1</a>.
</div>
<div id="ref-DiffBind2012" class="csl-entry">
Ross-Innes, Caryn S., Rory Stark, Andrew E. Teschendorff, Kelly A. Holmes, H. Raza Ali, Mark J. Dunning, Gordon D. Brown, et al. 2012. <span>“Differential Oestrogen Receptor Binding Is Associated with Clinical Outcome in Breast Cancer.”</span> <em>Nature</em> 481: –4. <a href="http://www.nature.com/nature/journal/v481/n7381/full/nature10730.html">http://www.nature.com/nature/journal/v481/n7381/full/nature10730.html</a>.
</div>
<div id="ref-tidyverse2019" class="csl-entry">
Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy D’Agostino McGowan, Romain François, Garrett Grolemund, et al. 2019. <span>“Welcome to the <span class="nocase">tidyverse</span>.”</span> <em>Journal of Open Source Software</em> 4 (43): 1686. <a href="https://doi.org/10.21105/joss.01686">https://doi.org/10.21105/joss.01686</a>.
</div>
<div id="ref-Wilson2019-ln" class="csl-entry">
Wilson, Daniel J. 2019. <span>“The Harmonic Mean <em>p</em>-Value for Combining Dependent Tests.”</span> <em>Proc. Natl. Acad. Sci. U. S. A.</em> 116 (4): 1195–1200.
</div>
<div id="ref-Zhang2008-ms" class="csl-entry">
Zhang, Yong, Tao Liu, Clifford A Meyer, Jérôme Eeckhoute, David S Johnson, Bradley E Bernstein, Chad Nusbaum, et al. 2008. <span>“Model-Based Analysis of <span>ChIP-Seq</span> (<span>MACS</span>).”</span> <em>Genome Biol.</em> 9 (9): R137.
</div>
</div>
</div>
<div id="appendix-appendix" class="section level1 unnumbered">
<h1>Appendix</h1>
</div>
<div id="session-info" class="section level1" number="6">
<h1><span class="header-section-number">A</span> Session info</h1>
<pre><code>## R version 4.3.1 (2023-06-16)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.6 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0 
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0
## 
## locale:
##  [1] LC_CTYPE=en_AU.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_AU.UTF-8        LC_COLLATE=en_AU.UTF-8    
##  [5] LC_MONETARY=en_AU.UTF-8    LC_MESSAGES=en_AU.UTF-8   
##  [7] LC_PAPER=en_AU.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_AU.UTF-8 LC_IDENTIFICATION=C       
## 
## time zone: Australia/Adelaide
## tzcode source: system (glibc)
## 
## attached base packages:
## [1] stats4    stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
##  [1] patchwork_1.1.2             rtracklayer_1.60.0         
##  [3] edgeR_3.42.4                limma_3.56.1               
##  [5] csaw_1.34.0                 Rsamtools_2.16.0           
##  [7] Biostrings_2.68.1           XVector_0.40.0             
##  [9] extraChIPs_1.5.8            SummarizedExperiment_1.30.1
## [11] Biobase_2.60.0              MatrixGenerics_1.12.0      
## [13] matrixStats_1.0.0           ggside_0.2.2               
## [15] BiocParallel_1.34.2         plyranges_1.20.0           
## [17] GenomicRanges_1.52.0        GenomeInfoDb_1.36.0        
## [19] IRanges_2.34.0              S4Vectors_0.38.1           
## [21] BiocGenerics_0.46.0         glue_1.6.2                 
## [23] lubridate_1.9.2             forcats_1.0.0              
## [25] stringr_1.5.0               dplyr_1.1.2                
## [27] purrr_1.0.1                 readr_2.1.4                
## [29] tidyr_1.3.0                 tibble_3.2.1               
## [31] ggplot2_3.4.2               tidyverse_2.0.0            
## [33] BiocStyle_2.28.0           
## 
## loaded via a namespace (and not attached):
##   [1] splines_4.3.1              BiocIO_1.10.0             
##   [3] bitops_1.0-7               filelock_1.0.2            
##   [5] polyclip_1.10-4            XML_3.99-0.14             
##   [7] rpart_4.1.19               lifecycle_1.0.3           
##   [9] doParallel_1.0.17          lattice_0.21-8            
##  [11] ensembldb_2.24.0           MASS_7.3-60               
##  [13] backports_1.4.1            magrittr_2.0.3            
##  [15] Hmisc_5.1-0                sass_0.4.6                
##  [17] rmarkdown_2.22             jquerylib_0.1.4           
##  [19] yaml_2.3.7                 metapod_1.8.0             
##  [21] Gviz_1.44.0                DBI_1.1.3                 
##  [23] RColorBrewer_1.1-3         zlibbioc_1.46.0           
##  [25] AnnotationFilter_1.24.0    biovizBase_1.48.0         
##  [27] RCurl_1.98-1.12            nnet_7.3-19               
##  [29] tweenr_2.0.2               VariantAnnotation_1.46.0  
##  [31] rappdirs_0.3.3             circlize_0.4.15           
##  [33] GenomeInfoDbData_1.2.10    ggrepel_0.9.3             
##  [35] codetools_0.2-19           DelayedArray_0.26.3       
##  [37] xml2_1.3.4                 ggforce_0.4.1             
##  [39] tidyselect_1.2.0           shape_1.4.6               
##  [41] futile.logger_1.4.3        farver_2.1.1              
##  [43] ComplexUpset_1.3.3         BiocFileCache_2.8.0       
##  [45] base64enc_0.1-3            GenomicAlignments_1.36.0  
##  [47] jsonlite_1.8.4             GetoptLong_1.0.5          
##  [49] Formula_1.2-5              iterators_1.0.14          
##  [51] foreach_1.5.2              tools_4.3.1               
##  [53] progress_1.2.2             Rcpp_1.0.10               
##  [55] gridExtra_2.3              mgcv_1.8-42               
##  [57] xfun_0.39                  withr_2.5.0               
##  [59] formatR_1.14               BiocManager_1.30.20       
##  [61] fastmap_1.1.1              latticeExtra_0.6-30       
##  [63] fansi_1.0.4                digest_0.6.31             
##  [65] timechange_0.2.0           R6_2.5.1                  
##  [67] colorspace_2.1-0           jpeg_0.1-10               
##  [69] dichromat_2.0-0.1          biomaRt_2.56.0            
##  [71] RSQLite_2.3.1              utf8_1.2.3                
##  [73] generics_0.1.3             data.table_1.14.8         
##  [75] prettyunits_1.1.1          InteractionSet_1.28.0     
##  [77] httr_1.4.6                 htmlwidgets_1.6.2         
##  [79] S4Arrays_1.0.4             pkgconfig_2.0.3           
##  [81] gtable_0.3.3               blob_1.2.4                
##  [83] ComplexHeatmap_2.16.0      htmltools_0.5.5           
##  [85] bookdown_0.34              ProtGenerics_1.32.0       
##  [87] clue_0.3-64                scales_1.2.1              
##  [89] png_0.1-8                  knitr_1.43                
##  [91] lambda.r_1.2.4             rstudioapi_0.14           
##  [93] tzdb_0.4.0                 rjson_0.2.21              
##  [95] nlme_3.1-162               checkmate_2.2.0           
##  [97] curl_5.0.0                 cachem_1.0.8              
##  [99] GlobalOptions_0.1.2        parallel_4.3.1            
## [101] foreign_0.8-84             AnnotationDbi_1.62.1      
## [103] restfulr_0.0.15            pillar_1.9.0              
## [105] grid_4.3.1                 vctrs_0.6.2               
## [107] dbplyr_2.3.2               cluster_2.1.4             
## [109] htmlTable_2.4.1            evaluate_0.21             
## [111] VennDiagram_1.7.3          EnrichedHeatmap_1.30.0    
## [113] GenomicFeatures_1.52.0     cli_3.6.1                 
## [115] locfit_1.5-9.7             compiler_4.3.1            
## [117] futile.options_1.0.1       rlang_1.1.1               
## [119] crayon_1.5.2               labeling_0.4.2            
## [121] interp_1.1-4               stringi_1.7.12            
## [123] deldir_1.0-9               munsell_0.5.0             
## [125] lazyeval_0.2.2             Matrix_1.5-4.1            
## [127] BSgenome_1.68.0            hms_1.1.3                 
## [129] bit64_4.0.5                KEGGREST_1.40.0           
## [131] highr_0.10                 igraph_1.4.3              
## [133] broom_1.0.4                memoise_2.0.1             
## [135] bslib_0.4.2                bit_4.0.5                 
## [137] GenomicInteractions_1.34.0</code></pre>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": {
      styles: {
        ".MathJax_Display": {
           "text-align": "center",
           padding: "0px 150px 0px 65px",
           margin: "0px 0px 0.5em"
        },
        "@media screen and (max-width: 991px)": {
            ".MathJax_Display": {
               "text-align": "center",
               padding: "0 0 0 0"
            }
         }
      }
    }
  });
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<script type="text/javascript">
$(document).ready(function ()  {
  
  // Map "enter" keypress to the same action as a cursor click
  function navigateLink(e) {
    if (e.key === "Enter") {
      $(this).trigger("click");
    }
  }

  var toc_items = document.querySelectorAll(".tocify-item");
  for (var i = 0; i < toc_items.length; i++) {
    // The link role tells screen readers this is for navigation
    toc_items.item(i).setAttribute("role", "link");
    // tabindex = 0 allows selection via keyboard tab presses
    toc_items.item(i).setAttribute("tabindex", "0");
    // Listen for "Enter" keypress when item is selected
    toc_items.item(i).addEventListener("keydown", navigateLink);
  }
});
</script>

</body>
</html>
