<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2023-07-10" />

<title>Bioc2023: extraChIPs Package Demonstration</title>

<script src="index_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="index_files/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="index_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="index_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="index_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
</style>
<script src="index_files/navigation-1.1/tabsets.js"></script>
<link href="index_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="index_files/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
  p.abstract{
    text-align: center;
    font-weight: bold;
  }
  div.abstract{
    margin: auto;
    width: 90%;
  }
</style>


<style type="text/css">
/* for pandoc --citeproc since 2.11 */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="/home/steviep/R/x86_64-pc-linux-gnu-library/4.3/BiocStyle/resources/html/bioconductor.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 828px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




<script>
function toggle_visibility(id1) {
  var e = document.getElementById(id1);
  e.style.display = ((e.style.display!="none") ? "none" : "block");
}
</script>

</head>

<body>


<div class="container-fluid main-container">




<div id="header">



<h1 class="title toc-ignore">Bioc2023: extraChIPs Package Demonstration</h1>
<p class="author-name">Stevie Pederson<span class="affil-mark">1,2,3*</span></p>
<p class="author-affiliation"><span class="affil-mark">1</span>Black Ochre Data Laboratories, Telethon Kids Institute, Adelaide, Australia<br><span class="affil-mark">2</span>Dame Roma Mitchell Cancer Researc Laboratories, University of Adelaide<br><span class="affil-mark">3</span>John Curtin School of Medical Research, Australian National University</p>
<p class="author-email"><span class="affil-mark">*</span><a href="mailto:stephen.pederson.au@gmail.com">stephen.pederson.au@gmail.com</a></p>
<h4 class="date">10 July 2023</h4>
<h4 class="abstract">Abstract</h4>
<p>extraChIPs: Package Demo</p>
<h4 class="package">Package</h4>
<p>extraChIPs 1.5.7</p>

</div>

<h1>Contents</h1>
<div id="TOC">
<ul>
<li><a href="#introduction" id="toc-introduction"><span class="toc-section-number">1</span> Introduction</a>
<ul>
<li><a href="#package-introduction" id="toc-package-introduction"><span class="toc-section-number">1.1</span> Package Introduction</a></li>
<li><a href="#data-and-workshop-setup" id="toc-data-and-workshop-setup"><span class="toc-section-number">1.2</span> Data and Workshop Setup</a></li>
</ul></li>
<li><a href="#workshop-material" id="toc-workshop-material"><span class="toc-section-number">2</span> Workshop Material</a>
<ul>
<li><a href="#core-infrastructure" id="toc-core-infrastructure"><span class="toc-section-number">2.1</span> Core Infrastructure</a>
<ul>
<li><a href="#mc-functions" id="toc-mc-functions"><span class="toc-section-number">2.1.1</span> <code>*MC()</code> Functions</a></li>
<li><a href="#tibble-coercion" id="toc-tibble-coercion"><span class="toc-section-number">2.1.2</span> <code>tibble</code> Coercion</a></li>
</ul></li>
<li><a href="#working-with-peaks" id="toc-working-with-peaks"><span class="toc-section-number">2.2</span> Working With Peaks</a>
<ul>
<li><a href="#import-and-visualisation" id="toc-import-and-visualisation"><span class="toc-section-number">2.2.1</span> Import and Visualisation</a></li>
</ul></li>
<li><a href="#consensus-peaks" id="toc-consensus-peaks"><span class="toc-section-number">2.3</span> Consensus Peaks</a></li>
<li><a href="#differential-signal-detection" id="toc-differential-signal-detection"><span class="toc-section-number">2.4</span> Differential Signal Detection</a>
<ul>
<li><a href="#using-fixed-width-windows" id="toc-using-fixed-width-windows"><span class="toc-section-number">2.4.1</span> Using Fixed-Width Windows</a></li>
<li><a href="#using-sliding-windows" id="toc-using-sliding-windows"><span class="toc-section-number">2.4.2</span> Using Sliding Windows</a></li>
</ul></li>
<li><a href="#comparing-both-sets-of-results" id="toc-comparing-both-sets-of-results"><span class="toc-section-number">2.5</span> Comparing Both Sets Of Results</a></li>
</ul></li>
<li><a href="#references" id="toc-references">References</a></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li><a href="#session-info" id="toc-session-info"><span class="toc-section-number">A</span> Session info</a></li>
</ul>
</div>

<div id="introduction" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<div id="package-introduction" class="section level2" number="1.1">
<h2><span class="header-section-number">1.1</span> Package Introduction</h2>
<p><code>extraChIPs</code> was developed to provide key infrastructure for a larger ChIP-Seq workflow comparing binding dynamics across multiple ChIP targets.
The focus is on:</p>
<ul>
<li>Differential Signal Detection</li>
<li>Comparison between ChIP targets or sets of results</li>
<li>Working with GRanges objects containing important values in the <code>mcols()</code> element</li>
<li>Visualisation of Data and Results</li>
</ul>
<p>The package was developed to utilise <em>existing</em> S4 Object Classes and behave in a <em>tidy-friendly manner</em>, whilst not being strictly designed under tidy programming guidelines.
Some functions may be analogous to those in existing packages, but the aim is to be simple and intuitive to use, whilst simplifying high-performance analyses.
Previous work by the authors of DiffBind <span class="citation">(<a href="#ref-DiffBind2012">Ross-Innes et al. 2012</a>)</span> and csaw <span class="citation">(<a href="#ref-csaw2016">Lun and Smyth 2016</a>)</span> has been instrumental in the package design, along with the tidyverse <span class="citation">(<a href="#ref-tidyverse2019">Wickham et al. 2019</a>)</span> and edgeR <span class="citation">(<a href="#ref-edgeRQLF2016">Chen, Lun, and Smyth 2016</a>)</span>.</p>
</div>
<div id="data-and-workshop-setup" class="section level2" number="1.2">
<h2><span class="header-section-number">1.2</span> Data and Workshop Setup</h2>
<p>Today’s workshop will focus on the <code>devel</code> version of the package, which can be installed using</p>
<pre class="r"><code>BiocManager::install(
  &quot;extraChIPs&quot;, version = &quot;devel&quot;, update = FALSE, ask = FALSE
)</code></pre>
<p>Alternatively, this can be installed directly from github</p>
<pre class="r"><code>BiocManager::install(
  &quot;smped/extraChIPs&quot;, ref = &quot;devel&quot;, update = FALSE, ask = FALSE
)</code></pre>
<p>Additional packages required can be installed as follows:</p>
<pre class="r"><code>pkg &lt;- c(&quot;rtracklayer&quot;, &quot;tidyverse&quot;, &quot;glue&quot;, &quot;plyranges&quot;)
BiocManager::install(pkgs = pkg, update = FALSE, ask = FALSE)</code></pre>
<pre class="r"><code>library(tidyverse)
library(glue)
library(plyranges)
library(extraChIPs)</code></pre>
<p>Data can be obtained from <a href="">here</a> and should be placed in a folder called <code>data</code> within your current working directory.
This data has been preprocessed <a href="https://github.com/smped/PRJNA509779">here</a> using the prepareChIPs workflow <span class="citation">(<a href="#ref-pederson_2023">Pederson 2023</a>)</span>, and subset to a region on Chromosome 10 for easier working in today’s context.</p>
<p>The two targets in this dataset are the Estrogen Receptor (ER<span class="math inline">\(\alpha\)</span>) and the histone mark H3K27ac.
Data has been taken from ZR-75-1 cells cultured in Estrogen (E2) or Estrogen + Dihydrotestosterone (E2DHT), and was first published by researchers from the Dame Roma Mitchell Cancer Research Laboratories <span class="citation">(<a href="#ref-Hickey2021-mz">Hickey et al. 2021</a>)</span></p>
<p>The sample sheet for the ER<span class="math inline">\(\alpha\)</span> samples can be setup by copying the following code.
As can be seen, we have three samples from the E2 (baseline) group with three more treated by the addition of DHT.</p>
<pre class="r"><code>treat_levels &lt;- c(&quot;E2&quot;, &quot;E2DHT&quot;)
samples &lt;- tibble(
  accession = paste0(&quot;SRR831518&quot;, seq(0, 5)),
  target = &quot;ER&quot;,
  treatment = factor(rep(treat_levels, each = 3), levels = treat_levels)
)
samples</code></pre>
<pre><code>## # A tibble: 6 × 3
##   accession  target treatment
##   &lt;chr&gt;      &lt;chr&gt;  &lt;fct&gt;    
## 1 SRR8315180 ER     E2       
## 2 SRR8315181 ER     E2       
## 3 SRR8315182 ER     E2       
## 4 SRR8315183 ER     E2DHT    
## 5 SRR8315184 ER     E2DHT    
## 6 SRR8315185 ER     E2DHT</code></pre>
<p>We’ll also setup colours for each treatment for consistent visualisations throughout the analysis</p>
<pre class="r"><code>treat_colours &lt;- setNames(c(&quot;steelblue&quot;, &quot;red3&quot;), treat_levels)</code></pre>
</div>
</div>
<div id="workshop-material" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Workshop Material</h1>
<div id="core-infrastructure" class="section level2" number="2.1">
<h2><span class="header-section-number">2.1</span> Core Infrastructure</h2>
<div id="mc-functions" class="section level3" number="2.1.1">
<h3><span class="header-section-number">2.1.1</span> <code>*MC()</code> Functions</h3>
<p>Many functions exist for combining sets of ranges, such as <code>reduce()</code>, <code>intersect()</code>, <code>setdiff()</code> etc, however these have been extended in <code>extraChIPs</code> to <code>reduceMC()</code>, <code>intersectMC()</code>, <code>setdiffMC()</code> and <code>unionMC()</code> as well as tidy-aligned functions <code>chopMC()</code> and <code>distinctMC()</code>.
This group of functions behaves identically to the core <code>GenomicRanges</code> versions, but retains data in the <code>mcols</code> element.
These are used internally by many functions within the package, but also have many stand-alone uses.</p>
<p>Define a simple set of ranges, with ids in the <code>mcols</code></p>
<pre class="r"><code>x &lt;- GRanges(c(&quot;chr1:1-10&quot;, &quot;chr1:6-12&quot;))
x$id &lt;- c(&quot;range1&quot;, &quot;range2&quot;)
x</code></pre>
<pre><code>## GRanges object with 2 ranges and 1 metadata column:
##       seqnames    ranges strand |          id
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##   [1]     chr1      1-10      * |      range1
##   [2]     chr1      6-12      * |      range2
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>When calling <code>reduce()</code> we lose all <code>mcols</code> information.</p>
<pre class="r"><code>reduce(x)</code></pre>
<pre><code>## GRanges object with 1 range and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1      1-12      *
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>These are retained when calling <code>reduceMC</code></p>
<pre class="r"><code>reduceMC(x)</code></pre>
<pre><code>## GRanges object with 1 range and 1 metadata column:
##       seqnames    ranges strand |              id
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;CharacterList&gt;
##   [1]     chr1      1-12      * |   range1,range2
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>All other functions behave in a simlar manner</p>
<pre class="r"><code>y &lt;- GRanges(c(&quot;chr1:11-15&quot;))
intersect(x, y)</code></pre>
<pre><code>## GRanges object with 1 range and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1     11-12      *
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<pre class="r"><code>intersectMC(x, y)</code></pre>
<pre><code>## GRanges object with 1 range and 1 metadata column:
##       seqnames    ranges strand |          id
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##   [1]     chr1     11-12      * |      range2
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<pre class="r"><code>setdiff(x, y)</code></pre>
<pre><code>## GRanges object with 1 range and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1      1-10      *
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<pre class="r"><code>setdiffMC(x, y)</code></pre>
<pre><code>## GRanges object with 1 range and 1 metadata column:
##       seqnames    ranges strand |              id
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;CharacterList&gt;
##   [1]     chr1      1-10      * |   range1,range2
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<pre class="r"><code>union(x, y)</code></pre>
<pre><code>## GRanges object with 1 range and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1      1-15      *
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<pre class="r"><code>unionMC(x, y)</code></pre>
<pre><code>## GRanges object with 1 range and 1 metadata column:
##       seqnames    ranges strand |              id
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;CharacterList&gt;
##   [1]     chr1      1-15      * |   range1,range2
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="tibble-coercion" class="section level3" number="2.1.2">
<h3><span class="header-section-number">2.1.2</span> <code>tibble</code> Coercion</h3>
<p>As this package is also designed to play well with the <code>tidyverse</code> the functions <code>as_tibble()</code> and <code>colToRanges()</code> have been introduced.
When coercing to a <code>tibble</code>, the <code>GRanges</code> element is coerced to a character vector.</p>
<pre class="r"><code>tbl &lt;- as_tibble(x)
tbl</code></pre>
<pre><code>## # A tibble: 2 × 2
##   range     id    
##   &lt;chr&gt;     &lt;chr&gt; 
## 1 chr1:1-10 range1
## 2 chr1:6-12 range2</code></pre>
<p>The tibble can be returned to the more conventional set of columns by setting <code>rangeAsChar = FALSE</code></p>
<pre class="r"><code>as_tibble(x, rangeAsChar = FALSE)</code></pre>
<pre><code>## # A tibble: 2 × 6
##   seqnames start   end width strand id    
##   &lt;fct&gt;    &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;fct&gt;  &lt;chr&gt; 
## 1 chr1         1    10    10 *      range1
## 2 chr1         6    12     7 *      range2</code></pre>
<p>Additional methods have been implemented for easy coercion of <code>DataFrame</code>, <code>GInteractions</code>, <code>Seqinfo</code>, <code>SummarzedExperiment</code> and <code>TopTags</code> objects</p>
<p>Coercion is implemented from <code>tibble</code> to <code>GRanges</code> using <code>colToRanges()</code></p>
<pre class="r"><code>colToRanges(tbl, var = &quot;range&quot;)</code></pre>
<pre><code>## GRanges object with 2 ranges and 1 metadata column:
##       seqnames    ranges strand |          id
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##   [1]     chr1      1-10      * |      range1
##   [2]     chr1      6-12      * |      range2
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>Whilst the <code>seqinfo</code> is clearly lost in the reverse coercion, <code>colToRanges()</code> also allows for moving a <code>GRanges</code> object within the <code>mcols</code> element to be set as a new set of ranges.
In this case, the <code>seqinfo</code> will be retained.</p>
</div>
</div>
<div id="working-with-peaks" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> Working With Peaks</h2>
<div id="import-and-visualisation" class="section level3" number="2.2.1">
<h3><span class="header-section-number">2.2.1</span> Import and Visualisation</h3>
<p>Broad and Narrow Peak files as output by peak callers such as <code>macs2 callpeak</code> <span class="citation">(<a href="#ref-Zhang2008-ms">Zhang et al. 2008</a>)</span> can be imported using <code>importPeaks()</code>.
This will always return a <code>GRangesList</code> for simple downstream analysis.
If importing <code>narrowPeak</code> files, the peak centre can be automatically added.</p>
<pre class="r"><code>fl &lt;- glue(&quot;data/{samples$accession}_peaks.narrowPeak&quot;)
peaks &lt;- importPeaks(fl, centre = TRUE)
names(peaks) &lt;- str_remove_all(names(peaks), &quot;_peaks.narrowPeak&quot;)</code></pre>
<p>(Names can also be set using <code>glue</code> syntax, however the above is generally more easily comprehended at first glance.)</p>
<p>One of the first tasks is usually to check how similar our replicates are, which is easily performed using <code>plotOverlaps()</code></p>
<pre class="r"><code>plotOverlaps(peaks, set_col = treat_colours[samples$treatment])</code></pre>
<div class="figure"><span style="display:block;" id="fig:plot-overlaps"></span>
<img src="index_files/figure-html/plot-overlaps-1.png" alt="Overlapping peaks between all replicates" width="100%" />
<p class="caption">
Figure 1: <span class="caption-title">Overlapping peaks between all replicates</span><br>
</p>
</div>
<p><code>plotOverlaps()</code> calls the plotting functions provided by <code>ComplexUpset::upset()</code> <span class="citation">(<a href="#ref-ComplexUpset2020">Krassowski 2020</a>)</span> and as such, can handle any additional parameters which are understood by this function.</p>
<pre class="r"><code>plotOverlaps(
  peaks, .sort_sets = FALSE, 
  set_col = treat_colours[samples$treatment], n_intersections = 10
)</code></pre>
<div class="figure"><span style="display:block;" id="fig:plot-overlaps2"></span>
<img src="index_files/figure-html/plot-overlaps2-1.png" alt="Overlapping peaks, showing only intersections with 10 or more members and retaining the original sample ordering" width="100%" />
<p class="caption">
Figure 2: <span class="caption-title">Overlapping peaks, showing only intersections with 10 or more members and retaining the original sample ordering</span><br>
</p>
</div>
</div>
</div>
<div id="consensus-peaks" class="section level2" number="2.3">
<h2><span class="header-section-number">2.3</span> Consensus Peaks</h2>
<p>The next step might be to form a set of consensus peaks for each treatment.
Just forming a set of consensus peaks for the E2 samples can be done manually</p>
<pre class="r"><code>makeConsensus(peaks[1:3])</code></pre>
<pre><code>## GRanges object with 247 ranges and 4 metadata columns:
##         seqnames            ranges strand | SRR8315180 SRR8315181 SRR8315182
##            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;logical&gt;  &lt;logical&gt;  &lt;logical&gt;
##     [1]    chr10 43048195-43048529      * |       TRUE       TRUE       TRUE
##     [2]    chr10 43521739-43522260      * |       TRUE       TRUE       TRUE
##     [3]    chr10 43540042-43540390      * |       TRUE      FALSE       TRUE
##     [4]    chr10 43606238-43606573      * |       TRUE       TRUE       TRUE
##     [5]    chr10 43851214-43851989      * |      FALSE       TRUE       TRUE
##     ...      ...               ...    ... .        ...        ...        ...
##   [243]    chr10 99168353-99168649      * |       TRUE       TRUE       TRUE
##   [244]    chr10 99207868-99208156      * |      FALSE       TRUE       TRUE
##   [245]    chr10 99326100-99326363      * |      FALSE      FALSE       TRUE
##   [246]    chr10 99331363-99331730      * |       TRUE       TRUE       TRUE
##   [247]    chr10 99621632-99621961      * |      FALSE       TRUE       TRUE
##                 n
##         &lt;numeric&gt;
##     [1]         3
##     [2]         3
##     [3]         2
##     [4]         3
##     [5]         2
##     ...       ...
##   [243]         3
##   [244]         2
##   [245]         1
##   [246]         3
##   [247]         2
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>In reality, we may prefer to only retain peaks found in 2/3 samples and we can do this by setting <code>p = 2/3</code>.
We can also also <code>makeConsensus</code> to retain one or more columns from the original peaks using <code>var =</code></p>
<pre class="r"><code>makeConsensus(peaks[1:3], p = 2/3, var = &quot;centre&quot;)</code></pre>
<pre><code>## GRanges object with 167 ranges and 5 metadata columns:
##         seqnames            ranges strand |                     centre
##            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |              &lt;NumericList&gt;
##     [1]    chr10 43048195-43048529      * | 43048341,43048357,43048389
##     [2]    chr10 43521739-43522260      * | 43522008,43522012,43522039
##     [3]    chr10 43540042-43540390      * |          43540298,43540245
##     [4]    chr10 43606238-43606573      * | 43606418,43606408,43606421
##     [5]    chr10 43851214-43851989      * |          43851672,43851766
##     ...      ...               ...    ... .                        ...
##   [163]    chr10 99096784-99097428      * | 99097259,99097253,99097249
##   [164]    chr10 99168353-99168649      * | 99168506,99168486,99168513
##   [165]    chr10 99207868-99208156      * |          99207995,99208002
##   [166]    chr10 99331363-99331730      * | 99331621,99331585,99331580
##   [167]    chr10 99621632-99621961      * |          99621809,99621828
##         SRR8315180 SRR8315181 SRR8315182         n
##          &lt;logical&gt;  &lt;logical&gt;  &lt;logical&gt; &lt;numeric&gt;
##     [1]       TRUE       TRUE       TRUE         3
##     [2]       TRUE       TRUE       TRUE         3
##     [3]       TRUE      FALSE       TRUE         2
##     [4]       TRUE       TRUE       TRUE         3
##     [5]      FALSE       TRUE       TRUE         2
##     ...        ...        ...        ...       ...
##   [163]       TRUE       TRUE       TRUE         3
##   [164]       TRUE       TRUE       TRUE         3
##   [165]      FALSE       TRUE       TRUE         2
##   [166]       TRUE       TRUE       TRUE         3
##   [167]      FALSE       TRUE       TRUE         2
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>By default, this takes the union of the ranges called as peaks in each sample.
We may instead prefer to just retain regions <em>covered</em> in <span class="math inline">\(\geq\)</span> 2/3 samples.
Not that whilst we have the same number of ranges here, the ranges are tighter.</p>
<pre class="r"><code>makeConsensus(peaks[1:3], p = 2/3, method = &quot;coverage&quot;)</code></pre>
<pre><code>## GRanges object with 167 ranges and 4 metadata columns:
##         seqnames            ranges strand | SRR8315180 SRR8315181 SRR8315182
##            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;logical&gt;  &lt;logical&gt;  &lt;logical&gt;
##     [1]    chr10 43048224-43048519      * |       TRUE       TRUE       TRUE
##     [2]    chr10 43521773-43522218      * |       TRUE       TRUE       TRUE
##     [3]    chr10 43540200-43540384      * |       TRUE      FALSE       TRUE
##     [4]    chr10 43606264-43606559      * |       TRUE       TRUE       TRUE
##     [5]    chr10 43851570-43851940      * |      FALSE       TRUE       TRUE
##     ...      ...               ...    ... .        ...        ...        ...
##   [163]    chr10 99097038-99097398      * |       TRUE       TRUE       TRUE
##   [164]    chr10 99168367-99168608      * |       TRUE       TRUE       TRUE
##   [165]    chr10 99207908-99208139      * |      FALSE       TRUE       TRUE
##   [166]    chr10 99331412-99331687      * |       TRUE       TRUE       TRUE
##   [167]    chr10 99621674-99621944      * |      FALSE       TRUE       TRUE
##                 n
##         &lt;numeric&gt;
##     [1]         3
##     [2]         3
##     [3]         2
##     [4]         3
##     [5]         2
##     ...       ...
##   [163]         3
##   [164]         3
##   [165]         2
##   [166]         3
##   [167]         2
##   -------
##   seqinfo: 1 sequence from an unspecified genome</code></pre>
<p>In order to make consensus peaks for each condition, we can split by treatment group, then use a simple <code>lapply()</code> trick, to work wth both treatment groups together.</p>
<pre class="r"><code>cons_peaks &lt;- peaks %&gt;% 
  split(f = samples$treatment) %&gt;% 
  lapply(makeConsensus, p = 2/3, var = &quot;centre&quot;) %&gt;% 
  lapply(select, centre) %&gt;% 
  lapply(mutate, centre = vapply(centre, mean, numeric(1))) %&gt;% 
  GRangesList()</code></pre>
<p>A lot happened in that code chunk, se let’s break it down</p>
<ol style="list-style-type: decimal">
<li>We split samples based on the treatment (<code>split(f = samples$treatment)</code>), then</li>
<li>Identified consensus peaks within each treatment found in 2/3 samples, retaining the peak centre (<code>lapply(makeConsensus, p = 2/3, var = "centre")</code>)</li>
<li>Subset the mcols to only retain the peak centre (<code>lapply(select, centre)</code>).</li>
<li>The previous step returned centre estimates as a <code>NumericList</code>, so we took the mean of these for each consensus peak</li>
<li>We coerced to a <code>GRangeList</code></li>
</ol>
<p>Now we can compare our two treatment conditions using <code>plotOverlaps()</code> again, which will return a Venn Diagram when passing a <code>GRangesList</code> of length 2.</p>
<pre class="r"><code>plotOverlaps(cons_peaks, set_col = treat_colours)</code></pre>
<div class="figure"><span style="display:block;" id="fig:plot-cons-overlap"></span>
<img src="index_files/figure-html/plot-cons-overlap-1.png" alt="Overlaps betwen consensus peaks specific to each treatment group" width="100%" />
<p class="caption">
Figure 3: <span class="caption-title">Overlaps betwen consensus peaks specific to each treatment group</span><br>
</p>
</div>
<p>Finally, we can form a set of “union” peaks, defined as those within a consensus peak for either treatment group.</p>
<pre class="r"><code>union_peaks &lt;- makeConsensus(cons_peaks, var = &quot;centre&quot;) %&gt;% 
  mutate(centre = vapply(centre, mean, numeric(1)) %&gt;% round(0))</code></pre>
</div>
<div id="differential-signal-detection" class="section level2" number="2.4">
<h2><span class="header-section-number">2.4</span> Differential Signal Detection</h2>
<div id="using-fixed-width-windows" class="section level3" number="2.4.1">
<h3><span class="header-section-number">2.4.1</span> Using Fixed-Width Windows</h3>
<p>Now that we’ve defined our set of peaks present under either condition, we can use these to perform a differential signal analysis.
Existing packages such as <code>DiffBind</code> use fixed-width regions to minimise bias from variable sized regions.
Using <code>extraChIPs</code> we can define these ranges using our peak centres that we’ve carried forward.</p>
<pre class="r"><code>centred_peaks &lt;- union_peaks %&gt;% 
  mutate(centre = paste0(seqnames, &quot;:&quot;, centre)) %&gt;% 
  colToRanges(var = &quot;centre&quot;) %&gt;% 
  resize(width = 400, fix = &quot;center&quot;) %&gt;% 
  granges()</code></pre>
<p>In the above, we’ve 1) created a new range from the centre-points, then 2) moved it to being the primary range using <code>colToRanges()</code>.
From there, we’ve 3) resized to a standard width of 400bp then 4) discarded all <code>mcols</code>.
We can now use these ranges to count alignments across our set of samples.</p>
</div>
<div id="using-sliding-windows" class="section level3" number="2.4.2">
<h3><span class="header-section-number">2.4.2</span> Using Sliding Windows</h3>
</div>
</div>
<div id="comparing-both-sets-of-results" class="section level2" number="2.5">
<h2><span class="header-section-number">2.5</span> Comparing Both Sets Of Results</h2>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1>References</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-edgeRQLF2016" class="csl-entry">
Chen, Yunshun, Aaron A T Lun, and Gordon K Smyth. 2016. <span>“From Reads to Genes to Pathways: Differential Expression Analysis of RNA-Seq Experiments Using Rsubread and the edgeR Quasi-Likelihood Pipeline.”</span> <em>F1000Research</em> 5: 1438. <a href="https://doi.org/10.12688/f1000research.8987.2">https://doi.org/10.12688/f1000research.8987.2</a>.
</div>
<div id="ref-Hickey2021-mz" class="csl-entry">
Hickey, Theresa E, Luke A Selth, Kee Ming Chia, Geraldine Laven-Law, Heloisa H Milioli, Daniel Roden, Shalini Jindal, et al. 2021. <span>“The Androgen Receptor Is a Tumor Suppressor in Estrogen Receptor-Positive Breast Cancer.”</span> <em>Nat. Med.</em> 27 (2): 310–20.
</div>
<div id="ref-ComplexUpset2020" class="csl-entry">
Krassowski, Michał. 2020. <span>“ComplexUpset.”</span> <a href="https://doi.org/10.5281/zenodo.3700590">https://doi.org/10.5281/zenodo.3700590</a>.
</div>
<div id="ref-csaw2016" class="csl-entry">
Lun, Aaron T L, and Gordon K Smyth. 2016. <span>“Csaw: A <span>B</span>ioconductor Package for Differential Binding Analysis of <span>C</span>h<span>I</span><span>P</span>-Seq Data Using Sliding Windows.”</span> <em>Nucleic Acids Res.</em> 44 (5): e45.
</div>
<div id="ref-pederson_2023" class="csl-entry">
Pederson, Stevie. 2023. <span>“prepareChIPs:”</span> <a href="https://doi.org/10.48546/WORKFLOWHUB.WORKFLOW.528.1">https://doi.org/10.48546/WORKFLOWHUB.WORKFLOW.528.1</a>.
</div>
<div id="ref-DiffBind2012" class="csl-entry">
Ross-Innes, Caryn S., Rory Stark, Andrew E. Teschendorff, Kelly A. Holmes, H. Raza Ali, Mark J. Dunning, Gordon D. Brown, et al. 2012. <span>“Differential Oestrogen Receptor Binding Is Associated with Clinical Outcome in Breast Cancer.”</span> <em>Nature</em> 481: –4. <a href="http://www.nature.com/nature/journal/v481/n7381/full/nature10730.html">http://www.nature.com/nature/journal/v481/n7381/full/nature10730.html</a>.
</div>
<div id="ref-tidyverse2019" class="csl-entry">
Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy D’Agostino McGowan, Romain François, Garrett Grolemund, et al. 2019. <span>“Welcome to the <span class="nocase">tidyverse</span>.”</span> <em>Journal of Open Source Software</em> 4 (43): 1686. <a href="https://doi.org/10.21105/joss.01686">https://doi.org/10.21105/joss.01686</a>.
</div>
<div id="ref-Zhang2008-ms" class="csl-entry">
Zhang, Yong, Tao Liu, Clifford A Meyer, Jérôme Eeckhoute, David S Johnson, Bradley E Bernstein, Chad Nusbaum, et al. 2008. <span>“Model-Based Analysis of <span>ChIP-Seq</span> (<span>MACS</span>).”</span> <em>Genome Biol.</em> 9 (9): R137.
</div>
</div>
</div>
<div id="appendix-appendix" class="section level1 unnumbered">
<h1>Appendix</h1>
</div>
<div id="session-info" class="section level1" number="3">
<h1><span class="header-section-number">A</span> Session info</h1>
<pre><code>## R version 4.3.1 (2023-06-16)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.6 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0 
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0
## 
## locale:
##  [1] LC_CTYPE=en_AU.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_AU.UTF-8        LC_COLLATE=en_AU.UTF-8    
##  [5] LC_MONETARY=en_AU.UTF-8    LC_MESSAGES=en_AU.UTF-8   
##  [7] LC_PAPER=en_AU.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_AU.UTF-8 LC_IDENTIFICATION=C       
## 
## time zone: Australia/Adelaide
## tzcode source: system (glibc)
## 
## attached base packages:
## [1] stats4    stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
##  [1] extraChIPs_1.5.7            SummarizedExperiment_1.30.1
##  [3] Biobase_2.60.0              MatrixGenerics_1.12.0      
##  [5] matrixStats_1.0.0           ggside_0.2.2               
##  [7] BiocParallel_1.34.2         plyranges_1.20.0           
##  [9] GenomicRanges_1.52.0        GenomeInfoDb_1.36.0        
## [11] IRanges_2.34.0              S4Vectors_0.38.1           
## [13] BiocGenerics_0.46.0         glue_1.6.2                 
## [15] lubridate_1.9.2             forcats_1.0.0              
## [17] stringr_1.5.0               dplyr_1.1.2                
## [19] purrr_1.0.1                 readr_2.1.4                
## [21] tidyr_1.3.0                 tibble_3.2.1               
## [23] ggplot2_3.4.2               tidyverse_2.0.0            
## [25] BiocStyle_2.28.0           
## 
## loaded via a namespace (and not attached):
##   [1] BiocIO_1.10.0              bitops_1.0-7              
##   [3] filelock_1.0.2             polyclip_1.10-4           
##   [5] XML_3.99-0.14              rpart_4.1.19              
##   [7] lifecycle_1.0.3            edgeR_3.42.4              
##   [9] doParallel_1.0.17          lattice_0.21-8            
##  [11] ensembldb_2.24.0           MASS_7.3-60               
##  [13] backports_1.4.1            magrittr_2.0.3            
##  [15] limma_3.56.1               Hmisc_5.1-0               
##  [17] sass_0.4.6                 rmarkdown_2.22            
##  [19] jquerylib_0.1.4            yaml_2.3.7                
##  [21] metapod_1.8.0              Gviz_1.44.0               
##  [23] DBI_1.1.3                  RColorBrewer_1.1-3        
##  [25] zlibbioc_1.46.0            AnnotationFilter_1.24.0   
##  [27] biovizBase_1.48.0          RCurl_1.98-1.12           
##  [29] nnet_7.3-19                tweenr_2.0.2              
##  [31] VariantAnnotation_1.46.0   rappdirs_0.3.3            
##  [33] circlize_0.4.15            GenomeInfoDbData_1.2.10   
##  [35] ggrepel_0.9.3              codetools_0.2-19          
##  [37] DelayedArray_0.26.3        xml2_1.3.4                
##  [39] ggforce_0.4.1              tidyselect_1.2.0          
##  [41] shape_1.4.6                futile.logger_1.4.3       
##  [43] farver_2.1.1               ComplexUpset_1.3.3        
##  [45] BiocFileCache_2.8.0        base64enc_0.1-3           
##  [47] GenomicAlignments_1.36.0   jsonlite_1.8.4            
##  [49] GetoptLong_1.0.5           Formula_1.2-5             
##  [51] iterators_1.0.14           foreach_1.5.2             
##  [53] tools_4.3.1                progress_1.2.2            
##  [55] Rcpp_1.0.10                gridExtra_2.3             
##  [57] xfun_0.39                  withr_2.5.0               
##  [59] formatR_1.14               BiocManager_1.30.20       
##  [61] fastmap_1.1.1              latticeExtra_0.6-30       
##  [63] fansi_1.0.4                digest_0.6.31             
##  [65] timechange_0.2.0           R6_2.5.1                  
##  [67] colorspace_2.1-0           jpeg_0.1-10               
##  [69] dichromat_2.0-0.1          biomaRt_2.56.0            
##  [71] RSQLite_2.3.1              utf8_1.2.3                
##  [73] generics_0.1.3             data.table_1.14.8         
##  [75] rtracklayer_1.60.0         prettyunits_1.1.1         
##  [77] InteractionSet_1.28.0      httr_1.4.6                
##  [79] htmlwidgets_1.6.2          S4Arrays_1.0.4            
##  [81] pkgconfig_2.0.3            gtable_0.3.3              
##  [83] blob_1.2.4                 ComplexHeatmap_2.16.0     
##  [85] XVector_0.40.0             htmltools_0.5.5           
##  [87] bookdown_0.34              ProtGenerics_1.32.0       
##  [89] clue_0.3-64                scales_1.2.1              
##  [91] png_0.1-8                  knitr_1.43                
##  [93] lambda.r_1.2.4             rstudioapi_0.14           
##  [95] tzdb_0.4.0                 rjson_0.2.21              
##  [97] checkmate_2.2.0            curl_5.0.0                
##  [99] cachem_1.0.8               GlobalOptions_0.1.2       
## [101] parallel_4.3.1             foreign_0.8-84            
## [103] AnnotationDbi_1.62.1       restfulr_0.0.15           
## [105] pillar_1.9.0               grid_4.3.1                
## [107] vctrs_0.6.2                dbplyr_2.3.2              
## [109] cluster_2.1.4              htmlTable_2.4.1           
## [111] evaluate_0.21              VennDiagram_1.7.3         
## [113] EnrichedHeatmap_1.30.0     GenomicFeatures_1.52.0    
## [115] cli_3.6.1                  locfit_1.5-9.7            
## [117] compiler_4.3.1             futile.options_1.0.1      
## [119] Rsamtools_2.16.0           rlang_1.1.1               
## [121] crayon_1.5.2               labeling_0.4.2            
## [123] interp_1.1-4               stringi_1.7.12            
## [125] deldir_1.0-9               munsell_0.5.0             
## [127] Biostrings_2.68.1          lazyeval_0.2.2            
## [129] csaw_1.34.0                Matrix_1.5-4.1            
## [131] BSgenome_1.68.0            hms_1.1.3                 
## [133] patchwork_1.1.2            bit64_4.0.5               
## [135] KEGGREST_1.40.0            highr_0.10                
## [137] igraph_1.4.3               broom_1.0.4               
## [139] memoise_2.0.1              bslib_0.4.2               
## [141] bit_4.0.5                  GenomicInteractions_1.34.0</code></pre>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": {
      styles: {
        ".MathJax_Display": {
           "text-align": "center",
           padding: "0px 150px 0px 65px",
           margin: "0px 0px 0.5em"
        },
        "@media screen and (max-width: 991px)": {
            ".MathJax_Display": {
               "text-align": "center",
               padding: "0 0 0 0"
            }
         }
      }
    }
  });
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<script type="text/javascript">
$(document).ready(function ()  {
  
  // Map "enter" keypress to the same action as a cursor click
  function navigateLink(e) {
    if (e.key === "Enter") {
      $(this).trigger("click");
    }
  }

  var toc_items = document.querySelectorAll(".tocify-item");
  for (var i = 0; i < toc_items.length; i++) {
    // The link role tells screen readers this is for navigation
    toc_items.item(i).setAttribute("role", "link");
    // tabindex = 0 allows selection via keyboard tab presses
    toc_items.item(i).setAttribute("tabindex", "0");
    // Listen for "Enter" keypress when item is selected
    toc_items.item(i).addEventListener("keydown", navigateLink);
  }
});
</script>

</body>
</html>
